generate-password on branch develop sends multiple messages for Apply password

## Overview
The base setting is KeePassXC being accessed by [keepassxc-cryptomator](https://github.com/purejava/keepassxc-cryptomator).

keepassxc-cryptomator sends the `generate-password` request and waits for the KeePassXC password generator popup to be closed, either by pressing `close` or `Apply password`.

## Steps to Reproduce
Easiest way is to use a JUnit test contained in keepassxc-cryptomator:
```java
@Test
public void shouldHaveNoErrors() throws InterruptedException {
    log.info("Please enter a name for the connection in the pop-up within 10 seconds");
    assertTrue(kpa.connect());
    assertFalse(kpa.associate());
    Thread.sleep(10000L); // give me 10 seconds to enter a associate id
    kpa.generatePassword();
    assertTrue(kpa.lockDatabase());
}
```

## Expected Behavior
Confirming `Apply password` sends back one message containing the password.

## Actual Behavior
`Apply password` results in multiple `generate-password` messages being sent back from KeePassXC to keepassxc-cryptomator.

Logging on the keepassxc-cryptomator side is as follows (association not shown here):

```
[main] DEBUG org.keepassxc.Connection - MessagePublisher started
[main] TRACE org.keepassxc.LinuxMacConnection - Sending message: {"action":"change-public-keys","clientID":"HXkLwnQ7wFpM5SwCrvnbwZXBEBJ2L410","publicKey":"IlJ5tSMuU2I7NCxqncWAvveq0v3nQ+woOyNK9duWWEg=","nonce":"MK3mKal5m0XjTgM0uW+S3shkZezyA3om"}
[pool-1-thread-1] TRACE org.keepassxc.LinuxMacConnection - Reading message: {"action":"change-public-keys","nonce":"Ma3mKal5m0XjTgM0uW+S3shkZezyA3om","publicKey":"ZtVZkyDmiPCNbJlzs0nKE7DBoYhjN0y4CuxGY4Rq9Ag=","success":"true","version":"2.7.0-snapshot"}
[pool-1-thread-2] TRACE org.keepassxc.Connection - Retrieved from queue: {"success":"true","action":"change-public-keys","publicKey":"ZtVZkyDmiPCNbJlzs0nKE7DBoYhjN0y4CuxGY4Rq9Ag=","nonce":"Ma3mKal5m0XjTgM0uW+S3shkZezyA3om","version":"2.7.0-snapshot"}
[main] TRACE org.keepassxc.Connection - Send - encrypting the following message: {"action":"generate-password","clientID":"HXkLwnQ7wFpM5SwCrvnbwZXBEBJ2L410"}
[main] TRACE org.keepassxc.LinuxMacConnection - Sending message: {"action":"generate-password","clientID":"HXkLwnQ7wFpM5SwCrvnbwZXBEBJ2L410","message":"qhhti1eH6I1bIyi+qbF7310QTTaNJQnk+XMuS9Sx3N4CSAJA2dvll+wUAypbU5jD00Novu/Fq9FtqqD9cdb06fA1rkZVr7oATq4eIruGKYkig1iMrel5JYCwQeY=","nonce":"/bKDHxbAklMKhei1+QHOVD1hmJYB7wTk"}
[pool-1-thread-1] TRACE org.keepassxc.LinuxMacConnection - Reading message: {}
[pool-1-thread-2] TRACE org.keepassxc.Connection - KeePassXC send an empty response: {}
[pool-3-thread-1] DEBUG org.purejava.KeepassProxyAccess - Attempting to save credentials
[pool-3-thread-1] DEBUG org.purejava.KeepassProxyAccess - Credentials saved
[pool-1-thread-1] TRACE org.keepassxc.LinuxMacConnection - Reading message: {"action":"generate-password","message":"n4aSiH6byrgeZVixT6TnQpfD8unZ2C/SOmWgtkI9VRR3d73FMz0I40+ufFFV7DkuwQGDZ+uHvoeV4ipMrOsqhB/gl3gi+1Q2XPlppojpceruHPgjQPMZvB4OPi1lJ5xusEF+uNkZV75V8L8azyrogKQ1efBNEN4sKaKmU3S2uDaQtUkHhmtx9BOnonqY5tQ6UotM7YqZ10VdcyPgZTpIJ71tUH7w","nonce":"/rKDHxbAklMKhei1+QHOVD1hmJYB7wTk"}
[pool-1-thread-1] TRACE org.keepassxc.LinuxMacConnection - Reading message: {"action":"generate-password","message":"o+80N2eiBE80QpFqOSO7wBFTVb6LlZq9GNxU3Tu1AxlN91W3VqaNC8Ro0P7YBee1uEK7qWcskK9eEzHTy12MvRomrXzhy2NqlyHRUuM6IA86pyXLsxgo+ZsTH3PcSLlpXLqw7eGLslwnswof4RjJH1s/WCb5lLmO9CaqdTmLJSSbkhhd53Lz2/dTN+wT+l4pVfh0rgQkr/5GkvnIn3IKU1BdVlam","nonce":"/rKDHxbAklMKhei1+QHOVD1hmJYB7wTk"}
[pool-1-thread-1] TRACE org.keepassxc.LinuxMacConnection - Reading message: {"action":"generate-password","message":"47xqKdyW+24nyZF4epZxD5dbCNt7Z8KuIJu1kv8YocN8bghY/Xp9CIZgU2r8bsieGQ1pguG/ZKLijYgd8WAm5ppN9sJ91lY2kmUu3s3GoNo4q3f62thgiFNs82SZTRHmEYZRs6iYhZk3xxWNJqdkf19MwVhB3C8CwXWpcBVPka4SW6IbjzEUvrZK0SuEw+SMXOnNyrsaH60J3x6Q+USDAPbsuJXj","nonce":"/rKDHxbAklMKhei1+QHOVD1hmJYB7wTk"}{"action":"generate-password","message":"4Dnm5sdfDMvjgL2luRMeRjwrj/xTtv7n83+HTg3TcCHJ4XPawPkcCGqr+bf85jNClSgzYW1X7UkdRUcVgXJdUOhNx9mq71dC+oa8S3etIBRajU1oE2MSDp/65N7iALACxsGPrb+hHOpWeuXLP/lfZSQd2FfFcGWsoto7+ftJWjGHMMjvNHQGVE39eRk5yZ/SK/rjS4LNVuT8sJcid9UAnGP2fq5X","nonce":"/rKDHxbAklMKhei1+QHOVD1hmJYB7wTk"}
[pool-1-thread-1] TRACE org.keepassxc.LinuxMacConnection - Reading message: {"action":"generate-password","error":"Action cancelled or denied","errorCode":"6"}
[pool-1-thread-2] TRACE org.keepassxc.Connection - Retrieved from queue: {"action":"generate-password","message":"n4aSiH6byrgeZVixT6TnQpfD8unZ2C/SOmWgtkI9VRR3d73FMz0I40+ufFFV7DkuwQGDZ+uHvoeV4ipMrOsqhB/gl3gi+1Q2XPlppojpceruHPgjQPMZvB4OPi1lJ5xusEF+uNkZV75V8L8azyrogKQ1efBNEN4sKaKmU3S2uDaQtUkHhmtx9BOnonqY5tQ6UotM7YqZ10VdcyPgZTpIJ71tUH7w","nonce":"/rKDHxbAklMKhei1+QHOVD1hmJYB7wTk"}
[main] INFO org.purejava.KeepassProxyAccess - org.purejava.KeepassProxyAccessException: Error: message could not be decrypted
```

I added `qDebug` statements after line 96 in `BrowserHost` to see what is sent back to keepassxc-cryptomator:
https://github.com/keepassxreboot/keepassxc/blob/9b7e54947b3ea6faf7066ab6401ad85c68c6d2b8/src/browser/BrowserHost.cpp#L94-L97

```
Sending to client:
"{\"action\":\"change-public-keys\",\"nonce\":\"Ma3mKal5m0XjTgM0uW+S3shkZezyA3om\",\"publicKey\":\"ZtVZkyDmiPCNbJlzs0nKE7DBoYhjN0y4CuxGY4Rq9Ag=\",\"success\":\"true\",\"version\":\"2.7.0-snapshot\"}"
Couldn't load passphrase wordlist.
Sending to client:
"{}"
Sending to client:
"{\"action\":\"generate-password\",\"message\":\"n4aSiH6byrgeZVixT6TnQpfD8unZ2C/SOmWgtkI9VRR3d73FMz0I40+ufFFV7DkuwQGDZ+uHvoeV4ipMrOsqhB/gl3gi+1Q2XPlppojpceruHPgjQPMZvB4OPi1lJ5xusEF+uNkZV75V8L8azyrogKQ1efBNEN4sKaKmU3S2uDaQtUkHhmtx9BOnonqY5tQ6UotM7YqZ10VdcyPgZTpIJ71tUH7w\",\"nonce\":\"/rKDHxbAklMKhei1+QHOVD1hmJYB7wTk\"}"
Sending to client:
"{\"action\":\"generate-password\",\"message\":\"o+80N2eiBE80QpFqOSO7wBFTVb6LlZq9GNxU3Tu1AxlN91W3VqaNC8Ro0P7YBee1uEK7qWcskK9eEzHTy12MvRomrXzhy2NqlyHRUuM6IA86pyXLsxgo+ZsTH3PcSLlpXLqw7eGLslwnswof4RjJH1s/WCb5lLmO9CaqdTmLJSSbkhhd53Lz2/dTN+wT+l4pVfh0rgQkr/5GkvnIn3IKU1BdVlam\",\"nonce\":\"/rKDHxbAklMKhei1+QHOVD1hmJYB7wTk\"}"
Sending to client:
"{\"action\":\"generate-password\",\"message\":\"47xqKdyW+24nyZF4epZxD5dbCNt7Z8KuIJu1kv8YocN8bghY/Xp9CIZgU2r8bsieGQ1pguG/ZKLijYgd8WAm5ppN9sJ91lY2kmUu3s3GoNo4q3f62thgiFNs82SZTRHmEYZRs6iYhZk3xxWNJqdkf19MwVhB3C8CwXWpcBVPka4SW6IbjzEUvrZK0SuEw+SMXOnNyrsaH60J3x6Q+USDAPbsuJXj\",\"nonce\":\"/rKDHxbAklMKhei1+QHOVD1hmJYB7wTk\"}"
Sending to client:
"{\"action\":\"generate-password\",\"message\":\"4Dnm5sdfDMvjgL2luRMeRjwrj/xTtv7n83+HTg3TcCHJ4XPawPkcCGqr+bf85jNClSgzYW1X7UkdRUcVgXJdUOhNx9mq71dC+oa8S3etIBRajU1oE2MSDp/65N7iALACxsGPrb+hHOpWeuXLP/lfZSQd2FfFcGWsoto7+ftJWjGHMMjvNHQGVE39eRk5yZ/SK/rjS4LNVuT8sJcid9UAnGP2fq5X\",\"nonce\":\"/rKDHxbAklMKhei1+QHOVD1hmJYB7wTk\"}"
Sending to client:
"{\"action\":\"generate-password\",\"error\":\"Action cancelled or denied\",\"errorCode\":\"6\"}"
```

This differs from the keepassxc-browser protocol and makes it almost impossible to handle what is returned for `generate-password` by KeePassXC.

## Context
`close` does work fine.

`generate-password` on 2.6.6 does also work fine.

[NOTE]: # ( Paste debug info from Help → About here )
KeePassXC - 2.7.0-snapshot
Compiled from the `develop` branch.

keepassxc-cryptomator 1.0.1

The current implementation of `generate-password` is broken and I managed to get a repro.
This is with KeePassXC - origin/release/2.7.0 and the browser extension 1.7.11 loaded to and run in Chrome only, no keepassxc-cryptomator involved.

First setting is: Association done, freshly start KeePassXC and Chrome, navigate to accounts.google.com and trigger `generate-password`, then trigger `generate-password` a second time.

Issue: this works once, but the nonces get out of sync immediately. It cannot process the second call of `generate-password`.
```javascript
keepass.js:1105 Connecting to native messaging host org.keepassxc.keepassxc_browser
keepass.js:120 Got message: {"action":"change-public-keys","nonce":"hLH41lzemFzFM+Tk4g8aEnsbnyyRuuVY","publicKey":"2xItIVV9xo5oJvGDtvCd3ExHeo2+Uhas520BRH8sgnw=","success":"true","version":"2.7.0-snapshot"}
keepass.js:640 Server public key: 2xItIVV9xo5oJvGDtvCd3ExHeo2+Uhas520BRH8sgnw=
keepass.js:120 Got message: {"action":"get-databasehash","message":"AwBGVOJvc6YNcTFiDKeWt37Qwwt2qe/zcqNDHLamHlPP67rSp0J3oj0k/uU7n1h+UF/MHy4F6KRioH8WQLlY0c0e3SlWG/9m+qR1Ar81o9DsGiabRlRzAK1tLqmDrAd4pWdkGpD34ywAawmVnHk/29qql1w26vWa9nb2JsORZzVr+xjN9ppRYgUk5FyQEoUr5drAsFVSwLDBKG72XjdwvEt6f8kxotNac0SppjNMgRDiihhaJpcxVHq3eLOnEIfF9T7G1hudfYZ0/OHv4PGdG8O3ud2g8PD8qj8o7hgFquKWbnrmLC5RqnJ50W8y24WAuySi6wDJ/MI48ksBF7Yk3WQ0+dQMSrc7KSti3NNSywqxnQXZOIsfmR2lpGb1G24=","nonce":"PDJ+5iEk4HgunkaYkpge27igdN8OGgnn"}
keepass.js:120 Got message: {"action":"get-databasehash","message":"4gNy8wGokedLY0DBWDMBinD3U6lUbvjGPF6ZkinF0eFUACTKu7Io4ToefhhpukwR5PDJbvK85TWpYidnTxUQiARKylJDKfslZjPYPp9OmWIrkuVUSs3xdmwKNs77YPxcA1iGe9gFN1TsTIn2guLEORy8doqqHPWCbAYI55Nw7+P7k1ZMbdEHbQbsULuIpPVPBV6+fzi0r/bXggFGLPBGBehytMoJe6RD8dTqqCWHBIy7406lscr6kzhC4FouvDGfbOsthShZ5JWvRnF1ZXirD//KyEMR39hSqhRRjmPGFinwQkurkds8idf9nwRnWsIIiD9rKYKqQ2WEY+PH86f8a4Gcx/Cu6oqcv79h1AGKxVh8Fq0WJ4ymNuTNrpEo/v4=","nonce":"zA7bLIsum0JRyL3bgsRUm+UxVfDY1FLF"}
keepass.js:120 Got message: {"action":"test-associate","message":"FlNHP/z6WYPEUKH4EwKGbHYk3ZRmUzFgADUIigYVc4ai6+Lvic7SYEW/dnPfofpPjfJS0+NoyVosTs6LAc5GyvptQpTgLnuDfZnMXRfmA37X2aefSy7yuFelBwg9m9mfjCITr3kRJeVFE883BVvp208gZiI9agZcWmUAMA61YJTNZlXyrUsUMRpmINDmmHOT5mW6lKUVGKV0b2l2QwVSISieMSbdKObfWKDEAmT9zBpvAtR7TG1ZNhCFzDlgQKFVhQqFz9g1tcGH6RAR4GYC2IndaUeiZphN0xPJYQoE","nonce":"RPfn4jpZi6DIl8jSV0mJziXgRIRtEdJE"}
keepass.js:1308 Error: No content script available for this tab.
keepass.js:120 Got message: {"action":"get-databasehash","message":"K7T9RVN57vqx3XFaUrRDGVwRGpqljZkhDugC5vwJcud4JCx5OlNgPysbjugzCp+jvWNrS8cSThqEPFnbwGbggWvOoMJUhKtLq7G5TrMWWw2NgnNJRrTWqZ5dBb9jeQPdiykKR60/R8h1WOmOigIgJFuRnY8uU4O/dc0JJsx5VZpfo9uAM5qCWI3YFgViZZRd0S0BAJRz5DxnoEQE9hURmDzoQUvWDPRjuvjvNX7AFMeNmfEqTPbD5EA2f92/KBRiHZRKGE74uubOTdVcyazi6G7L63CCvIz9uo/kjvgiZ8A4ATjZY50vFzI83YOo568rMnJvjSV+ZnWE7ewSJisC3rRJ3J8nai7eEwp0UCcRpZsy4mLLU/uVe5SVzdyGK5Y=","nonce":"Z4LCxUWN7C7Z5cKDsZ9Gn5nXSmTJFTp5"}
keepass.js:120 Got message: {"action":"test-associate","message":"uGOz1XD6zLz6CIFwlO0YArYGeTP88UpORCKi5dqXJd204o4uS3bmtiowGuHdzskOjUHPB5KhHj6OoqX+DUk5CNM/56KCXvHQa2VszNA1KK1BlKQMlbc0eEkNA7SaRcpRBl1A16Q4cHKJr2Oy2+uFd5gu5S867EBCKT9EJcdnDDqkInFjS83IPmgTWUi4M22MDqXpg5ND1FD19dNwiov8Y+TkDIRYKgKQnlXEXiG9qZzUHCWqrYAPukd1Pfkp7FLb3O/fEhRP3ajW9X6zHJNPQB9r/0Z41W8oBZtG9WY6","nonce":"qB1ngyJVAU1VyjG/IyFOe8pg5pQPhsj1"}
keepass.js:1308 Error: No content script available for this tab.
page.js:209 Cannot send activated_tab message:  {message: 'Could not establish connection. Receiving end does not exist.'}
keepass.js:120 Got message: {"action":"get-databasehash","message":"n84cmF+TuTyEEOhamttz3zv1o6YD7Doy1zacB89Qb7dEQ775J0aqk7NbVQFVmNnCRrjchz5GCPlAtPbJXBRetCk82oIDI+8/IqWBtEskaWbFD7v+dPIJwrx/zqkdQcimRFj53ijw4wQJwlK37Z+3pzZa25Y/s2auC3naHBe++1IDAahihzblFSxCip7MLEEkA46+R76mz2czQnmLu77pqALvF30H6ONmpC8DbywJ2UtJU7Gdt82WuB146XctYJbCbxKqy+cd6eD32JmdxTMizntMTM4V4w8qhe1mKBJRkRKzf6nDlnvkrf8ssvUI4zQfbnHfIE6WRItYy3qa8nArdK3fFq5Ha5fI6PyWd9Tde56V3yMWAuEBZap/iYD0tv0=","nonce":"JbXNYLBPXVlwjiVTbVzGGDWHl6GrN4/a"}
keepass.js:120 Got message: {"action":"get-logins","error":"No logins found","errorCode":"15"}
keepass.js:1188 Error 15: No logins found
2keepass.js:120 Got message: {"action":"get-databasehash","message":"ZzQ9uB1iaogCkrB4HT/gjz7uBt1M8kdEUCCOep5HKJsrMeMx0evunRmPY5akXE6jreJrTx23esoSTnejxYTtYVaGeM/X/tDpPMuNavvWeYaxKEn8aCJODyM0Y0ejc8t7bt9pwDV/EIC2c3tBgZmOTLx9dgJ3Woa+7d7705I6w4SnMPGtGE2Rg5IK/KfB7ctMu4gCBk0sSHKPep9lBalNrkFk+QxZMaHLRlCZXQ4xGqQ1KxJqknyObDXvW9zuPaoqou/k5/gL57KWVATteEg7FzetUfJzQVVgHUeuCOpI3vYCjZ5shhZdf3+h+TWg+2ade5pLExZnN3Y0SfOu9GC2eAx2uVnWkDaTnOOwcWv+H1RBS+UujsnJiyAtM5DDxKU=","nonce":"NGmEVlOXisQIqPoj8pM8ZjNHMyDSJoZV"}
3keepass.js:120 Got message: {"action":"get-databasehash","message":"QKiue1kLnKYwDrUshvnW1ru9D0VV2AIzrb0qCv+Uklu0zdTTNubNOPOxit64kj24YYGiVRZTJ+vP6CkyuadMSe2NMxRNwLvGLC3JdL0zkKEYfJWFTKBpGqHy6X0jiXQjDJNDSWtPutGJws5Sq4x9Fswgr4PZS4V/EjHY/5m1ektT59+0QhHlUBb0PSJMYazzGktOrOMEcZJ8vUPCJD+1lGgWojMzQ9UZg/6Q9XWpY6uWZQzh0myv2xbqlnmK+RUUIJlEjy3XZnKnK+OWsLhUrrO+SLKTUJeDzoonCgYh0Jr5NWP/bdMSlutEscn1YMOsJf6qyYPYS0P8u+CIETYMD2VjfAk8IiHcZFHiKiAcbXa+TV59fNMeFPZj8uF9IP8=","nonce":"gok2IDDFocIeo5FrxcKmvmuVXOVsXW/h"}
3keepass.js:120 Got message: {}
3keepass.js:120 Got message: {"action":"generate-password","message":"yJVWhESxhO6XyzPva9oDvWGAFbb/A3gqQnXSdjprMImiB93lHvcttVpEeaTGl04USDRG4oo9BFG+8y/5Xq7OIqwMmBtj5NvGYxLwzauzwN/1x/lKPl5KYEQK8A6aXzykbfovqpEnON2ZVsdH0lk2VOprB6U9BwbW3A7G5nkLmaP9pIKrnhlMMu4vFca56kWfnFdbv3b5/DGC2LiigkDA6yyvOAk=","nonce":"VoAZg9lwN+EwmONz3C7Hd9ttaA3CJH64"}
keepass.js:362 Retrieved password:  69$FuTe:gT#Sv+=R_3;X
5keepass.js:120 Got message: {"action":"get-databasehash","message":"D+UD9PYMaXDGxbPwBK3h3+OsohvRoYM/fC30Mc9uQg7+u+KhAApzC5l3OgH98LVEL0iyJGDB5ifnUxUUDhggreBhXAkCcb7Twrhi1nwD0I+frXm39Q2+ZT87cVqVoMK2OI4h7VZSG74lQxLa3wRh1R4xh0u5HY6viHGgNYNePsIqR9PZynP4jcHEnxYVI5nLDc2hI/JT9sWMhTlnO6gb7Md/1iK4AZ1+Mme7//pCRTYFP87l1ndEJv7PlAHf6YFlgu7Dkd09ZthQYjgV7rSV/YB7rQB800u8C49ySqZpWzj3V0OitWBs/GBPm+NVW4laJL6FuY4K7n9iFHKMn6Dxt2pVgCBO4nclwZ14uuX9OgmY0L1jpKhsk2omjBCZAYY=","nonce":"8tW5LxrSIaNcfuuR9MZBAxNsFRwVLiyl"}
keepass.js:1172 Error: Nonce compare failed
5keepass.js:120 Got message: {"action":"get-databasehash","message":"Eg3Xx8ua+fhl4jk7gllJ+6GsXJ8rV719gDkcb1k7dKXdZr06/88zY1lQpuo0qOpwjptWVU/34edbL8fcYkyiUFqie1wNcM3otjVjc+u+vVQFpUsfJPNYO0hitUg8CIkyhwqfC/ZPErboiqXDPLxs29q/0gHFKzqYtiXXjI285gded30RWqvBxIYaPTEuCT6fkxF6n91apbCS3ZubM909qOduVMOmVd8SIHTFqztayzydhUKpUGWn2UQmw3SoAVIOzvWJuEEKzzvGm+DDsB7sL3crIl84y8KEwEgIzoJLNVRYJEOs7FxUJ6NBGEEEX6B9HiNFi62QV/KJvofOGVY0xlfd8+pfv8JTlCEoMUv0Oo67ZOQ7MWeYm6F4nvUXb5U=","nonce":"3u6i/XZC7tLr9n74ZXSpSNJhttb4ot7m"}
keepass.js:1172 Error: Nonce compare failed
5keepass.js:120 Got message: {}
5keepass.js:120 Got message: {"action":"generate-password","message":"jFu5KlPDi6G+qsuPRLj4Qeoc8sdZ61Ftm9Pa4r5g1pz/h0vfiWMSFScfOjm/1euHuFiWMIqfnNmA/xBLYh1ZMNI3/rqvqYE6h8X52+Tza7z1kVCzFGKVeuREH+Ov4eIBG+1fzjqA7tSziXkJ5bvInQIBeI75c9nMTonpfIz52jOVZrdAGEwD7wHaSWRk1wjkBFs4U+xDTNG4Spi5pZYIRp39r9Q=","nonce":"VjYry40SuMz54N3LM4t6vPJgIy9ZbasD"}
keepass.js:1148 Error: Nonce compare failed
keepass.js:360 GeneratePassword rejected
keepass.js:362 Retrieved password:  undefined
```

This can also be verified by having the same setting, triggering `generate-password` once and clicking on the KeePassXC icon. The dialog shows, that it lost connection.

```javascript
keepass.js:1105 Connecting to native messaging host org.keepassxc.keepassxc_browser
keepass.js:120 Got message: {"action":"change-public-keys","nonce":"yDCpyfuj2UjEXYGJGq92G0O5T0A9lvbi","publicKey":"vYBnnPMkt1J6wBMAL9DmxPoTmZykd9ENsnyjXYSQAiY=","success":"true","version":"2.7.0-snapshot"}
keepass.js:640 Server public key: vYBnnPMkt1J6wBMAL9DmxPoTmZykd9ENsnyjXYSQAiY=
keepass.js:120 Got message: {"action":"get-databasehash","message":"rj1yabjymK6oTpmSAl0mqOVe1gXsYPOgwhy/oevomLyjaOhX58gJ8clAn661s9L/EeiN3C0wXtOvIFAwnejemBaCNoQV+JKAS+nd3OvmmeOufn/4Dl8IJEIBleHhdg6qed3Df2UzlY/ILIFgP+eei67mfLymJFpnt9JyuM36/JiA/Cn5nEKO/t/Xw8wpaz8VqsrpAdFS9I7wUpa0y5NmThzrxJ1TIQAW4JUsnx4moGIMEiN9kVfkesifX4s+p0urs1jGtlZo2Dx1APkYufS3Ikhxb9y6LRmCQSJfxamNoBkn9md8qsNwhR6sp+bf546E07GIVf3RMz03m3bQ9HbWkqH6oayMKludru0stR9Klj4qdRNEa1n4CjgzttfN1Kk=","nonce":"1mK97DrhA5bJROG6/fnyXUZAerT9BPvq"}
keepass.js:120 Got message: {"action":"get-databasehash","message":"G6ayjRiJ+Joz45lmQ5YWZdwAsXKmf8gzgfd7Wo97yz/VQ86ZaxQBQ/C+A+424tgcynQQDwP5+9Ye6s694m//usUgXiYiMyS4G8anMQojWIeE3LLqedOsMYR9BPFYWijUT77a4zNZS5T/g9ATMh6td2bSIcaIbzzT4MEowsRY3i8IDi6qdIPHEAv372QnBrtCvLDqvsDAwJYkfW/6vFCGvu5DTHaATPkZhs/GlyWDigcEVy/C+LYdTlVhGLFPnzYBjPF8v481Oo4G0rkaUHnKpvqEmGJRDc2qTkuK0UdErDLGSqALnNPeKlRpWFeLQw5QlHZM2EeITA8w4aUu0oUrI3dobpzhdOZJCDmbIZHgDHSa+gtY+TPCgU8GwMOPthA=","nonce":"06Jc+gkkDHR7HHMCWb+YkNtWV2E5dtwF"}
keepass.js:120 Got message: {"action":"test-associate","message":"6i4OB+P7Mx3u7khYDWrmsN22lPkv85Xk1q2a352UE4NnYeCVQwB2kHQv3xAMaWjxo0nbyW60sDo9Wrq89fXC+Jlo17VrQMDXr8sSz+VpmXh+0OjzfpoxFswPpFuT7IvY9ovB7SVWJf49qech2teyqeZacVF0mcHQ3m4ZIKNn+tSBSfn2DvWq6JjIP2ClR4oQy4XJTx6NbKg2gL9gDLl0y5QqI0pGqWEBNWmzcMC20GCftDYU+7cV3fFnLHPOOwARdHUvzcOAoV8/+zA8BKDPJNr698ILGzxDPHl2sXZt","nonce":"GUm30cCMDNNCBOCvX0NjSGBit3VRqnbA"}
keepass.js:1308 Error: No content script available for this tab.
keepass.js:120 Got message: {"action":"get-databasehash","message":"HOQfedctrnXTuBBnljZaaOuhsSJs2EH347K7wl49vSPcmaOC20pO8nVQaMNSZCFhrqy+v0p0oYHGaFkK5x1QWPyX9CH9eSIe/Ra3+wg4AbjewsWwta6n+3eVjt6Cy67Y/UY+h/CxqxQRUJmKCm5XoG0NIOoE+oFabZVQhvzvYfpNN0kz3XBLF2zCkyiRpCZjFuIcRSNQAduqsusiXku9MXiymBCUSc7akLtPDRy/rbUX/9CiuwS07e/xzGX9tO6vGDZbpTOnejuC+HUh8ldJqunGwbr4lxriasVqLGDLvpOWGtUyEiBNZTOIOdOMubEYaC4op8XzPOXTWmamBZQJpvygohkUs0yZWCSnlEOvkquXl0lFmJTyrqjm4zdl8kg=","nonce":"ctEahXyBBVkDChmfnLWcg/O8xioskO3i"}
keepass.js:120 Got message: {"action":"test-associate","message":"NJeqhqb92KPxRvR34ss1QYwvAC4AS3h/QWGu+nzpFixZQgo1w5wEckok+I64WA3sbEw5ldjPmLQqkLrb067bdPeLB6Lq3GPsI9C7AyV5E0+44Uw05eqJze1tcDGmOL2fZ/x8vKjEMu6bNZIsmH4XMLGwZQ6sS/uwOHVrmLBFGqERh531bvAujoK6vEfVAeu83w0t/PvopHx4URGkoAFwfpP6Lu8oXmsoEzuhaCKHB6R4qWbR8iC0Iv70vBxHAvDYekh5kjcLbXNCaR0FmnCkULkvkM/drRjj6xhacQO/","nonce":"hA8rCscSEaywHPwhHnURJrq8RQc+IlR5"}
keepass.js:1308 Error: No content script available for this tab.
keepass.js:120 Got message: {"action":"get-databasehash","message":"JQ1jR+0KP6cIhZ+eb2rNCxsmEqoZmRSVjHM2Fty3qmGvrJwyLR4bwGlYFbbqssCrcAR0QzcVPqHEXJqv9ReD0rIhrI7mncrinRf2Wmx7oN1+8KIV+moahaV5KmQxsLKGhKxLg2gSxFWS5DfrkPiS3ybJykGi9cTpjTAZpdgMTaLv2KfMqtz0llc+vvzFHv1jaoG1XkebvVtvyjcRr+q8bBf99V48iBd0RGLANKcm0WCEz4MgMj0CTbe55YCMWDEjP+/1H8k/OhtopJjecBTGKPmM2b2rmbvbMHnVb7g65k3jVtDSrJDPwy1JGQ1TMFXIgE670IlYHIJFiihrwxrr2IQ0IS6yZeMncSEJzEtNkOX7CVbijIyyRhGKuL9PYlo=","nonce":"0faB/XvoUYljwMYDtCjL1VA4+qA7apfT"}
keepass.js:120 Got message: {"action":"get-logins","error":"No logins found","errorCode":"15"}
keepass.js:1188 Error 15: No logins found
keepass.js:120 Got message: {"action":"get-databasehash","message":"2Fl7oaCew4nGi5q3ErLpYh7wRlLEOQhO10crdZ52HfYPyta9TbpzgJORFMCbYtNmmtTRtLL24w+pANrCxORZNzCUcsbMkVbIpEFt6T+U1itMZHi27MFaRORpPWOb0d26410FCqSApgznHZPV0YoH14bSTsjPGe2P3G+XQrJaLrRYDJCpcsscjqrUErdYtapeRRsbkZ5UJkbbJiRaMKMSavDpuahii5p8OVwSAKwzaekVZhHm61bBBZusZIOHAUptjgMHJJ3FVNWfJkKS1S2/Em9Lig8N8KEkuCdf5KKd2w2HU3Idw/59g4EI0h/fsNNtnZPz8MAppgfWKiGyXmkQ0sY/mLcFDgz8s4g4y7bZDWechoGDx3XsB/yEU0GJWW8=","nonce":"eDmWuEiClDnkRaFwM4NSTyJny4yPmKF5"}
keepass.js:120 Got message: {"action":"get-databasehash","message":"V9rWikzklAZl1OquBtjLhxoXCjiWqVsdRM9LSUfdwfb+66H8TyTsHijhisef50GjCR86fp0K7H//vYQlf2ufElMseU94jZbSCzVbdqFjAzknYOOrPvC4f+VjOcU7yCM8ZvkZboJ0TNpQCwJZ1LE+RVzWCBlsxFUTNgqOdyPE9uGEV4sqLchYiQgXn/QalxUhwMyuWTFJmPgYrsFLkYfKENMVkJy54nNey8dFZctC8IsNcNPPuG0+cLrJVZ2CuzBpJVLxYcwuhYg6nHbVW36ooWwlU4kgBRjk46dBJZh3KaMWTTa6Y7R7W2NT3mxcsIgRdK5/cYrBPf6lplwnBA8TdAvIETdWr8f0nrRx+xChe1yb2iS1zrtt1cJhSKMis2s=","nonce":"mz9kz+5Ne7zPguhYbBCncAUlfULe7dat"}
3keepass.js:120 Got message: {}
3keepass.js:120 Got message: {"action":"generate-password","message":"m7Z2u1gVNXjL6VTjwoKnARtqTXh27jxLa8PcehlnRTu1H0DSBcWECi7uJQsFE+Xy26fUZoUnIa/eZKxP0Xzb8XZjMzNWAzibkypBogT72iYyvMDx8u15mcO8V608FODaNl51sqNmw7ubJGTT8AGkMYUjUZhuAu9eAtfKF22z1lwKTCoPB97gnCzhGQzmM7yLJyS4QsE63ShVlJM8eg1bahRXslry","nonce":"op7s2ZBHbqReo3e920tC67JdRRfP6xFU"}
keepass.js:362 Retrieved password:  zi\*E/]}(B7]aD~{=_jV
keepass.js:1188 Error 5: Die Verbindung zu KeePassXC ist nicht möglich. Überprüfen Sie, ob die Browser-Integration in den KeePassXC-Einstellungen aktiviert ist.
```

A second non-working setting is as follows: run the first test as described above. After that, keep KeePassXC running, close Chrome, delete the association in KeePassXC, start Chrome again, do an association and trigger `generate-password`.

Issue: this time we get two `generate-password` responses, none of them can be decrypted.
```javascript
keepass.js:1105 Connecting to native messaging host org.keepassxc.keepassxc_browser
keepass.js:120 Got message: {"action":"change-public-keys","nonce":"WmwUPVVRZZf1FUC75MFDL3zae4hU0Zth","publicKey":"TqR1LGqBEvaUzlw0fsQJTuxXBWI6aKqEJbY2KZ38knE=","success":"true","version":"2.7.0-snapshot"}
keepass.js:640 Server public key: TqR1LGqBEvaUzlw0fsQJTuxXBWI6aKqEJbY2KZ38knE=
keepass.js:120 Got message: {"action":"get-databasehash","message":"GeQSk6eYEftjegFi9ePym/jEOBQwSWuSQ6aqT7XdJ2wC+VNYTQNNCL/hJ7gzM9+U2KROrMUTgXEITcQe2pNNwggWKt58Tot0ur4YLxVXg02thGCrUA4HWzSMKtaz1+X2cgGfcNoBS9V9VyEQkDj9sQoR4gyHoafRXmO8QNSvyzFQrSMkKFUq0KDdgox8tjaxwq0wfzjbLsK5kFwZN2uKrccbsLG+tG/DToE1ab4hqeOhdrZPW1g0cRRWtYB1TeGQt78/RI83lJDRr1gianUYSQXd5GiRw2eO9bh/p1riQBz+d7V7zJdRdfcoxr1EeajhFSvhvnT7H1JR4Cwwd16yumTvtRn3pJHMyK/ETuy3nEhjJ+kpEx5jtWh24TFuVCk=","nonce":"e7XmD9CQ40bJVqczfKpnE1ukiWDhrrVb"}
keepass.js:120 Got message: {"action":"get-databasehash","message":"ddaFW6/mL58JF3NuYpVOHHMZGQAusssaXfJAhVihMky7eyqY+g1VyzjITBx7jqcrfj6wklgKNkcrh+0xxde4pPC5/WgBKQO7K0QAJSsdGtbETk3qZIZVsuoW48zEqH5Zn7wB+8qLEKg9hQBfPt9rvLNGrUBl8mnVY4ynOvmn7DTG4bXVv/YiCzq/reKnRTDwRRgji1Qdsb9Bd6xeGSaCG4Na1Nevmz3P/4s/OWzPui8dzoerxRHcJUeJ+V9EIYH4pzMgbvcVaFDt3imro1sYEiycKg4xnhvlAa8W125YYj8qbYqWmsvFUt3enr+Wjsoe8dgOK6QdcSnBIB7csa+Oa25orwUfpbNfB7xDQRiNVOORqFykRanLIXLh2S58sC0=","nonce":"opXHd6KJw1KQ20af1vHYMsPDmDUCpoVe"}
keepass.js:120 Got message: {"action":"test-associate","error":"KeePassXC association failed, try again","errorCode":"8"}
keepass.js:1188 Error 8: KeePassXC association failed, try again
keepass.js:1308 Error: No content script available for this tab.
keepass.js:120 Got message: {"action":"get-databasehash","message":"S3lE1NYjPkpde5wzVhJQxkyiJ+dnU3CCH4VsQlGvZqXK/h1ITerJ5ZljvgVGVj+MXn6mMe1B9gQ8fZ3g/P7Cq7Z3n2MhJLcdvTwhJ4kgy+jMMq72gqJpaeLsZukInxjuzQERcQK6Pt8//rvW9wo2HNCH70ThpclNxHZEa2jXDYDPIRuFb7zyf3ChVJNHz62Vh092qdSrTwReDd8zgGFoU6ao++3L6Oj9fQ5NZimr6GSVX+PFmSGxe0v7NNgVt4beTNeNYDVEkuNw7Bwzks4QhB0KoietLvNYZEV5XoAAmLoEALYK0MHWsPX8o2jWRVEZXNoZYPLLOzncCpCNVHSSrjoCLQbbnssYKcvJL58RFFjlJWGf4UV8Wr0/P00FO5g=","nonce":"cgVsa/wcod7Zig+9tlkhrWcLJVRLvorU"}
keepass.js:120 Got message: {"action":"test-associate","error":"KeePassXC association failed, try again","errorCode":"8"}
keepass.js:1188 Error 8: KeePassXC association failed, try again
keepass.js:1308 Error: No content script available for this tab.
keepass.js:120 Got message: {"action":"get-databasehash","message":"mfu10uTRciAwT8T77mHw39SAQ1uAFUp67U2iECFKFOiKbDEqAoJys9/mpGi+dGqbJVAlGWxt9yPDMmO3lANiVwCH4F29HUNkK8GCxXx+7aFqOLj8zn5r34uoOkrUGkAFt2egO90Rzyij8FUiNXkZcg1N4Fr5hEs99wxqHp8BCa9PL7mmVXWSUVTh+QG+5wZpna7NQrCBNJSakDJINmvprLIu2QiLUh8Up+D6EKtwzpI93Yn6IBEfJ9sKMeKy6sXgFScLgCLDX1UbTTVconuhnoMFJd4EPrY+B/gFWen9VKz6ZzKCUzHktqDZECdKdz3SziPoR/MuWFKTU1AyK8Zm7SiYVjM0yIifhRZHyWaH7HMr4owPzdm6PEJkydRbovQ=","nonce":"Kxw9xVV432RW3Az8MzugKhRMp0nYpCtr"}
keepass.js:120 Got message: {"action":"test-associate","error":"KeePassXC association failed, try again","errorCode":"8"}
keepass.js:1188 Error 8: KeePassXC association failed, try again
keepass.js:120 Got message: {"action":"get-databasehash","message":"AAEa1tC+eJVKeVUfC3McX+5FNXMw3bD7YkJCIBgrcletnHpR8sRZXRQ7VFSD3IvuQ5R8uE21k81y28l893eiOYQ5CO+9KscKpBD6bk9KmqTxc3ng6DOBKiC/gTXnq1PRWtfWez5cKGSWbh4X0SqElzfNrwKNhND4EZ6tlFEMSePHhqbhbRSV4eIxQWzUrVGQVpULaic1YZFmcmoYd2+/u55ljGbTZnbwYi3sScaprQIzSnK/NDTCiAdjbMoElDdraTZhcCNG1/m0Efins2tWwLnlxtBjWVc3p6MSxRumB3U+tRsEuDr38Mu39tEzoV9yz6sKiwxIsbXSfivFPpp9LneSndXOX9R9X7DmWI+csm+Vw/k0f4ujv8H0us9yzOI=","nonce":"R1iQlcUPGDh9OK+IPybw2Xb/M4eJervQ"}
keepass.js:120 Got message: {"action":"associate","message":"dtTu6PZpvIaH4E2csfo9QE2CfpsmvEmDTPKByOo8gKTL6Ni2xCp4M0Z8yiMc60TxnoDM5eIhboFlGSI7sux0xAtxXjNxkz4AUUYxB+YiI4vsteEKcGyrvYqb5gXLIa2aWMLzXUeScDo5sF/vI3fZ+CYN0WRD7uVONsnBd7gqG4sym8Fz0AuLpn+GxWHWQAvSCtGlC6Hiu0k1DD0hvGQp0zPmViwk/f6N6v9nirsR4IGaRcgZg2UPOOnQ5RhHz9vZSKxQks4tHcMyAOlBSc6zT6vdkBgjFZyxtL/z+ZRpaA==","nonce":"L6IV8Z0WQB7iOtfG5BfgTxRAjZjRtYrA"}
keepass.js:120 Got message: {"action":"get-databasehash","message":"knGwWVGRJvUyK0rxNVz1jz7s06Km1ZF5vZ2wim8aV67lNkhLr4DdPAmy7QGO9OtOGd+YExQzktspars0ksXeqE50u8qCiu2RcLWECpPM3AtLMUvsJ2iwd/Ii0yqRLmHb4eHQqRydnsj9Ghu7fFjlpt2P+9X6u2WrTbldSD4SGo16K/EDY4CBnt6TVhqu/9UlGdtuqZngca5J57BWrr/FtInk2OIC1fff9RFxAl1Ps9enRmM0VUy9dC1xGjtyi2KuecNPE2VESb9JUWcF+WQxKpVhZ4U+Fp9Jl5euqrrX7JXXURoc5mm+RmqvSB0vUbOZ4luFuKeMHuhsqSA1oPGds23AbcsDG82xwuvaJHIkuuwJLLTitNGo05bHJd823p0=","nonce":"1AkjLdFrV222I/l3SwFfT5CEW7BiNC3K"}
keepass.js:120 Got message: {"action":"get-logins","error":"No logins found","errorCode":"15"}
keepass.js:1188 Error 15: No logins found
3keepass.js:120 Got message: {"action":"get-databasehash","message":"1i0CUUepnI4BkC3pYTugMJObpgTD3n9x8/qUsdE+LkD0ifGK+vTHE3eZj70PTx0jPAjhslUXR1YBSzbabEyJjovMo9VZEBe0xgPu8niDgMuxcQrNp1LodRVYV2TdOjDGaFaN3sqBRHM3gNGKve1lr+KOxoRuJ8IcVgq5MAT+EhB9EiWtiquz9ZV8S0HWWmvOEL3dQ/N1ou4Rua1fCJT75DqbjF3BWDUpz7sox3lrrmuvM6e/T76kRSW61ZOFgezrFwwB4wnCs/swPtp+G3GPIQa9Hh2EULQN1kJqKs2YjTqNbvwpEzbq6jvv1zpmBVXGlRT7eOwcqKqF6HzPSOobWT3zOeD2n1qIG8cwza3VR7NjUUJUJv/OI21J28L6kLc=","nonce":"hMHffprEkPGi7Mh9P+WvlnoJYFHXBQKH"}
3keepass.js:120 Got message: {"action":"get-databasehash","message":"8vkItACOPQ4+ofTOSoEnvIC0eBgi+Hiq/Xstl8sZfbvi/teZve07DZvcNnZJ3J5mCnUnWiXUgy4xCRErrftz0/XO3QNrU2hHOGeYrrnawqM2vcqna3vov3HMcyECVYkPL06dA2lQdqoH2Tt9AShg2d9thk64Q5JKSAIU84yZ7s8rps8pEDkQc4pLd5cv1FCbwccbiku5QPct6tBOdFtgDTpZ5betWbpC6AtpI8U0g3FbybSbyjcWlD5694haFA9mizh686OSOueK+LlPRtej2BzODHCKZ9bF42JbG2iWoss4LPzM8OeXeCH4h8A1NA3EfIP0SQJHRiMu93KLW5p5zYzFiZhoPb2A8X0MbwnR9pw9obQtROdpLlWKMkRQADM=","nonce":"jg/lycYY5To+/pZgRtyw7S5CkkZPEzhH"}
3keepass.js:120 Got message: {"action":"get-databasehash","message":"pTMMXpVm5hL09yqbyjNC5M3WfyNezU/lMewXCD1Vd+qVcHONarCzbsLWD9n9PT83XjpEDnPAFIK8eSnGASoDhlVqPUTE504ZArqm5u9w9urmJgDI2dDA1hRnvLK/AAZ52xZm5y60aAiLh7AkU9y/14O+QRNpSF0z5ZIeRTRI7wDyuyYHQKYTlBKW8cNEEJHWns0thbdh6X/08XTaAAMau3hKitkfjvnB7l8XY15U3Q4pGLBsLE2TmYGj3Xpu0wbQfGrqWHsTxITAthnYwq9iCY2nGyZJ+7vmcgrCiJRgLu9IkH/Rhtx/aEuEuhpQx2BH4KohpLwdmrjYU9JSooDfaEjpl61IFJxwcPqIWPJvYOe9ZNVOja1rCEu6Vdl1TCc=","nonce":"iAyTQMo0bF6FXvWvODkf50VjgZveTVlB"}
3keepass.js:120 Got message: {}
3keepass.js:120 Got message: {"action":"generate-password","message":"oNyjnsPKvxofAk0PO77rasB1I5aRuPBP6AR+OT1/UQX5YkoPIgpEcE99FiXi+Is67JzA6sfsbDPZP1vB+c1j7Qa9C3g56fTepjNFGWtRbfPORaYoAi4/1o2p1k90nvSjDg10Dl9ivPF+rrj18bC0JJaPY9IaKw/geo1DDFygn0ivfgGhdyj88HQem5VgNfNEfTdf4nNGWfMrDQt628PwCpZHkRbe","nonce":"eYDw/1syTigXGVuNpJNHygqpMI/q1mRE"}
keepass.js:1188 Error 4: Nachricht kann nicht entschlüsselt werden.
2keepass.js:120 Got message: {"action":"generate-password","message":"HUip2naXn0iarc5/oOTHPtdBUhxwUinA8dLtUsY/jSHAvPvblMilW3/yjGMWaqrFfTCtk1kj4afPSKqYTnvSbmOdcW6NDJtXxPlAoadk3kT4VqV2jfeyyT064LJf1rEadG2AuMHDzAP4oaNB7FO0c3lGiTewLKHJt2Oicmp9ofCdQDGHdsA2Q5sT3VkfPaCIOa8QEZGWYq8SP7rnJXEvAajwZSJV","nonce":"eYDw/1syTigXGVuNpJNHygqpMI/q1mRE"}
```

The icon saying there's no connection is different when I test it here. I used the same setup as described in the [comment above](https://github.com/keepassxreboot/keepassxc/issues/7358#issuecomment-1032281201).

The first call to `generate-password` is successful, but the nonces still get out of sync immediately. Surprisingly this does not affect keepassxc-cryptomator.

```javascript
keepass.js:1105 Connecting to native messaging host org.keepassxc.keepassxc_browser
keepass.js:1188 Error 7: Die Nachrichtenverschlüsselung ist fehlgeschlagen. Läuft KeePassXC?
3keepass.js:1308 Error: No content script available for this tab.
keepass.js:1188 Error 7: Die Nachrichtenverschlüsselung ist fehlgeschlagen. Läuft KeePassXC?
keepass.js:1308 Error: No content script available for this tab.
keepass.js:1105 Connecting to native messaging host org.keepassxc.keepassxc_browser
keepass.js:120 Got message: {"action":"change-public-keys","nonce":"t2kEBvszwJ0bI2RdA0q7s2Wsj5e4VS1l","publicKey":"mbbFKet5dNCMRcwpJgDFLq45M2Y6K4LO1ABShbw/OyA=","success":"true","version":"2.7.0-snapshot"}
keepass.js:640 Server public key: mbbFKet5dNCMRcwpJgDFLq45M2Y6K4LO1ABShbw/OyA=
keepass.js:120 Got message: {"action":"get-databasehash","message":"0mpJhstvU9xHkvYne6UGHsnai0OfhLRN6q9r2xAtmGkraxbsNpVbVeAU8XV7pAUbWuW+Js3JUgo2ttANGXVSqW3HgRyTsM+ezPy6Xi/ROleuK2233WasJ5w3rjkym9osMqW8guQSNmyCahNg82WB74A7sHzW9bBhn6m9NcyMS2IqJyvo8/Ekx1goc5H7aungLh2vqPEpL6Je0zxUaSSNRTkDPoWYhoYqnVOcOG+mq++R2dKy8zlCJ+nKGEKJRdigEwUFrQBxiA0wrAg5Bls2IRg6orkMYkR9uTuTUqEXyKmWbcLAC+LpyAM+6O+ve4XuV2jT/C7kzm9vOLtLL31jaomQXrrwKBBvy/ZegkHeB/1a5hf0C7iK8Ip1z4/D+OI=","nonce":"I/FQCjvF6P15kFLEK46fG9A59bnyxBoi"}
keepass.js:120 Got message: {"action":"get-databasehash","message":"+wDuLVIkHsdlUYRTGXRRTXwhEJ68CcKTFYUcd8+BQINNaKDSzYsUiuQx9AP7vkNDrLc4xvJuS4bUqoHgWtgwiBpLC+8Ifk5omL60/xwRR6eKJt/egQbUfW0QWqLMq/XyMzyLnCdt5k07DFVy9bng/qQZ/HF8nc+PP4/hYBFMFttE1kk2LKupwBF8yDRaaMulicgDFo+j/5Wtp8C7j7YLsbMhs6Jm33thg1d0BWVYs2SEvEJcVdjrPO2McLv/VRpOhdUDRsNWXKvIxNv+gKVgQy6mN9BOiDm4O/EuXgogs3b/rQ0rhyk4Cw8hE75JGOpzAqNrEJxehofwxW12jYoZScrH5VNriIuRDhkD/BBybna75lLVwCoa773DifqkLAA=","nonce":"fVLpqNQhcbkESEhRcAYnkvgf8pc4s++i"}
keepass.js:120 Got message: {"action":"test-associate","error":"KeePassXC association failed, try again","errorCode":"8"}
keepass.js:1188 Error 8: KeePassXC association failed, try again
event.js:93 {message: 'Could not establish connection. Receiving end does not exist.'}
keepass.js:1308 Error: No content script available for this tab.
keepass.js:120 Got message: {"action":"get-databasehash","message":"qXKR8hcG7c+ptxbGzg7SB28IMfLpoQHZ+cnzEuxLKLW+XFx216vbEx3Y1ZREQgBUt1v+vFqsVHsWAqGXqc2oT8vcbWHsbw4wCpYP2xTdW3t5/nqGn6MOQmNhy+ainc503yw9WiETU22ylH1U+Qk4OKwWkovOG70980BJN/l2//EEhuHS7BKmnKQeyGapyCxYphrLHvxvhf7hmOHNeOqxXX+RSJ10opcRIYf9nVJRpEA072w3vKpPHkz/TTfHV0J0iTjvH59NSSVTA0iyS0DoLgSwCCRDxmO4NUeFjXQMUQyDwFDzEnM/GQSbNZ2g27X39Ex1+Hl0xDRwthH319UwfQXaaCGpLXL/Zg++rzepkx5pCh9bRYLT9aDm+r+9Y3U=","nonce":"5VRibOUTCvQb6q7FdteSfhoUA2TeSnse"}
keepass.js:120 Got message: {"action":"associate","message":"h1PoHZ4AxGBfCUFEiJ3xnZJhj7uh5CF8nsjjbrXKZ4YgRCUzTvL91iAkX0tUXucOOQLsiPgp5wCEbeJxCMhYzEGkZUS4bplaJx5IJvX+y6LWL0AnLvGUB4e2bk/iL2RYOCJGbMIqjrTDMwFY2cwcfknY2jSfymZ/+/fqQ2XalpvSBUDXnoP06UxgXbKg4u2XFtSWuJGZK533gNlnH4YqBZCTqRUIOHBhlbtSM6dUWLt/r6NsqHmq1BPXZrws+Ewp3vza79jNIATsyPgzZ8CHzoJE87735T64wLeT","nonce":"VKmUVFp/e/NznF5XEkJVSOvN5uFpVp1s"}
page.js:209 Cannot send activated_tab message:  {message: 'Could not establish connection. Receiving end does not exist.'}
keepass.js:120 Got message: {"action":"get-databasehash","message":"RevuYcOB30Wn/MNInKt872v5VJNGauD+lf2u65nsa/W13WvQAQ0IJCqNbzfLGsmejqJviDzNGkNAZn4sUGF1gDkXuHEkedK3WoA6UROmDFjIWEd8GGEi+A2D0TMIMKmGPSkp+mkk2dTcSPv2rhpbiS5WUrDLXMNfBTAYuLy4KwQz+c6RFknVuCcJ4bpqZ8DUxE1hVXET//sgdxRwTHKllMKkKLPZmrYB8gxgYoC/BZZM+9nZib50bhqVyBkE8JXJ5pqiaD2dBnGpnlxbEzsO0iKjoeLGJ4XakPFaX6OlswYP3dGCUoHAb/iMpXTHkhwoOFjtIeipR0lnVQ8LrbeK+Ij/WerSWMx60Qb7goM7PT56v6i+I38hXAyTQK8Xxy4=","nonce":"n2Q9DumFHQmzwicg5x4m63qmruRzDekZ"}
keepass.js:120 Got message: {"action":"get-logins","error":"No logins found","errorCode":"15"}
keepass.js:1188 Error 15: No logins found
keepass.js:120 Got message: {"action":"get-databasehash","message":"QKgTI5P3tEm6UjtDaoSRJ55kLANcRq9StkBwURYgwaMfe76EwDDT6SEo6HyuSibdfKneYv5us02fzJsD7zCevArylIx2IADnOOAgD3C/q/f4izwnujIs3mAhL15Ndsd0xglp/c6ACfvKmDKiqACm45H80EzgjdkdO6eGa0KsXCJAoiNVT5dGIcRpJX9Vz41gmJZ7JbzHwWvOt3SPWL6LznGhrOJTLA+ZHZbWnp160ayHbKr+vNN5NQJlhgt5ALqWSkR+Qwgzy0LLKLIyeGyjQsc5c4tDDewfHjkBo5f4wbnkvAZ8qutHGHpgTSmpApc5g/RdZgWibtiqj8CWqgPUCvyxdPnESt82ki/yBx8A3wIkt5olGQ34FN3EzXGY/jM=","nonce":"yXBcdpXsZJWvenawSOYvKT92OJefZ+Tc"}
3keepass.js:120 Got message: {}
3keepass.js:120 Got message: {"action":"generate-password","message":"max/fPlpbisloLOBAL2kmfWvYVHrlhDayDEVnrHtHdCrETIDbnvaKgdWCiKOAtHDwnkYV1wImbV7ljt55gUcs6XZ3FqKSTjpRnhRi+6aVhW9LuL2rRvnEUKrPlMNXGaZReXF0NBvdZddXivS0GAs67BSNlp7Tb4ibGs9+iZfIrlv0HNnyWtGpDb3XWyMMJsR0pXQtLBG/qZA0/I6bGfgEomKieyK","nonce":"Rbx1gZgx8GQVZNPlJMMkndZ2/h9cXdLX"}
keepass.js:362 Retrieved password:  o#$,Ia;T"$4)[KJJB~/v
5keepass.js:120 Got message: {"action":"get-databasehash","message":"Ok/94B8Q2Bp9om1rtKP6l1SrTM1kqmjcqAMNH2H5ji41cXbvSSEeqA4uM+LYE1LG6jSQgU+5gs4uDkuCYcUpWG42cobhtUD+rCrIO+1ELwaiDGVc292zmgSJueAnIaht9G91SvnEyFrXFWP6lM72e/Z46T/6EE3lJqymDnbjveUOo8mL+5jqG8xMdXBaqBclCSxIv90l2Jm/4WoPzqjD+fW40cLKWKL4dc8+bdOudkxDTvEWTBTV4ChanFUuGpm48PPKNcKaADk3+eYxiLNjfhB7T6dL4rRle/mpf9fnWpehs9FZDLSRziA3TR3zmIstsOAHfSt2donhKRFdsbGgrlLyLNbPB04tUVnChCRSyaTq1+QoSYL8hy1H0+uDsJk=","nonce":"Wx6Wnb9bixMwrWUULDlA/ceMRW2L9cgs"}
keepass.js:1172 Error: Nonce compare failed
5keepass.js:120 Got message: {"action":"get-databasehash","message":"VCwjb0QR9YnzCFMQsutQLYW3wpvp+IodYoWBaIbzV8vYTPs6qYa7CPq14HiIXG6u2BenS8TwqMeQ4yJOPdTu2083TlSPVgYMxX6LPwX0QHLknIcUUPFqr9JhtjZeDbYWOfFqLtmwFp15pIytsZ6YZM8bu+MdaOeTuG3ONB/waSqx77aNN2LM6bDwrXt8TGrjYRodf27Imvz1CbAxmVSSXyRsBVRW/s6Inzdt7VXTcqhhNhz0gfNaDFznQLMqdvTYI/TnrsYsrV8Wj4q6BzfftUhbKgmZ/hMfZ/IymNcQ0i/iJnVzZdfnh6KCKB1q8TtIV1opXVqHkvsA2C8Z83kNHxiBAf/KDiUj6uC3B5ZTn/VrklIVeFWrjUd81KTrdU0=","nonce":"93NMbePhH4okboo4IMyg2L6zp05W6XRo"}
keepass.js:1172 Error: Nonce compare failed
```

@varjolintu Yes, the same here. I used a different database that has logins for the page for the same test. The nonces stay in sync.

EDIT: looking at the console tells something different tough.
I thought the nonces stay in sync, as the icon still tells there is a connection and offers the logins for the page, **but**:
```javascript
keepass.js:1105 Connecting to native messaging host org.keepassxc.keepassxc_browser
keepass.js:120 Got message: {"action":"change-public-keys","nonce":"S8GiXwlOXRL6++K2Z+8SuSgLqQVNULwB","publicKey":"u2AR0moGWm/MzFsY0fv0kuEb4Xe3Nmzqn+jg6gSRLUA=","success":"true","version":"2.7.0-snapshot"}
keepass.js:640 Server public key: u2AR0moGWm/MzFsY0fv0kuEb4Xe3Nmzqn+jg6gSRLUA=
keepass.js:120 Got message: {"action":"get-databasehash","message":"gsjeLGiJqX4Ahw3ngJNVBGVq57O68aXrihvhLT81GovXCDG8z2j/am58FYBeuOyZyOcG6Wu4wZzUZLJdfeBZWE2GeHfz0c5yVYkRGFev5J6875YEHhxc8ZU8+kYHWoNzq3DdM8CXeUZ6U2ejcDyhvkw+Bu/xea7zwolPi1vFAj71MeDkHNF53UzS2rUk/kNlsI60ovCyuJakwnC+aX72L1EChzmzl2uUYEhD1qhupuoaHYK3rm5POcOXnhXB0Dk5nNPjARkm4C0NvTk59cTN5nYGYDuRUVhA4V4uUYNKYlEhNplhZCIHWHh6uVrJC11unzs1DgVjAMxAwE+R9pLZSWxZ4fPZOB3JM8KDY5BvvUmpKVPUI8vtc0iDxo6SHXk=","nonce":"0Ow8BuVYv8CjBZGmST5FFRbqKPGpj9xy"}
keepass.js:120 Got message: {"action":"get-databasehash","message":"mr11s+CchhAS4pg2YdGGZQ3UZy0gBZJDAN297jP8zOFflbj+F/xzGxvE9LTnMqKBqOnKexyRmTU3nAKshHye92D3+4JtiuiMHRS4l/mSYuOKEgFrjIroSuiS+f/qm8lCL4NnCMeyr6bd3F9ttSO/mENfKq+jDRmj9xTwfqRagbd9qGOno5bqtnvnFekCFO7FmM9F+5RmQMQr+IGotqZ8HxLfK1mM73Ypv0q79R6/0C7DkHJw4pRAfKC3sBO/VZUj6kl7VqoFPKF5uJKaTMF4qEdmes93Qbxxb99mNobxysno5RHtTOL6GUFxlzCCnpJfMbps7NDDoNTjLTH5PgyMLdC4/z9+Qtz1NV5iHnteFYWx/8nsnXeX6XHX3W34bXA=","nonce":"CmBvVqV8sVDSoj/bch74E6OroAS4YiRq"}
keepass.js:120 Got message: {"action":"test-associate","message":"YRku13BrPZECYVgri9z5afxS9wcivVM+JaBteooDMZtMGbdhNJnPUzVEH2t6xIDgoI3qxF+OtUF031MfpfQBeURuPgybp4404LIEWF5KlHvsSLFBv1qGQcvv4ZsJtlBr1OF8siyZ4r8OF4v0ErahDCccS28AO0hkBhmxw8u5Fa+Uw38xW2CA4F1CARNeTAVXd8tFJKlscoDP4EcqMxwOhJmZwUpyv8LFAfVuU+XXdOwv3nAxsNtfeaIy6/eXa65Kjs0xBo+N6cPtEKUdNdxDI0lL511N8t60Gois","nonce":"DHd0puzdZonpTynlFYgq+CnxciT3jGT7"}
keepass.js:1308 Error: No content script available for this tab.
keepass.js:120 Got message: {"action":"get-databasehash","message":"Bxy/WS2SRoS5cQ2B40Jy9nRBH1gAGovJNWMi9A+0iz6WAtTxoW2RjPRHH59pn3gQR3F7XVjAPEDc8zj+xTUuIgzA3k8R616jmcDYt8XJfHLGm3n76S0x+6z3aTZ/ph9k/CIkAHlxRRJRKZKR/B/TJKVxXtN9/0tlafSZkAH9GSLTdYra09cvBr6aCNpsIC8zwfcJEIccEF3/20duccFIYkTY0z/WM4bKyDKfIOS24hartuwncH7wAZTeK68paJmHlwfkS0q0I2Hkw57JjehlvjjJC2eZXWAt3zvTcqZ/13Ya3mQo/BJC7jr7+21b3IuLvAk4rKtAkkyv4ySrPo1DOE8jc6VcwWsTAfaE9JilT2ZjObbR/CwWS3g7LhSxh6o=","nonce":"StNGYv6DmJC2gzK4V6yzMPjI1Tw5e3mO"}
keepass.js:120 Got message: {"action":"test-associate","message":"H9v2Gzn5Ygo/OAbLxtuOLsM7dg/tITRDstYdQp68dM7Ig9JuHWtOMxjUAm7rIY2/yX/HEyCiik3P5ibIzzVAl/2RHcFbM0JtTdjRTZbgiji7Kel3PKk+AehTWRP+ieU15x0cTCObP/LUH2iZt4gkqIrm6eWh9Y76TcYDxEptTDRamk94BVTpbdwpzOwr1yZKefJExEp9EpfvOuPsWPoWIda9YnNvKWdltglV2CP5umI9qD5U7TazjcBAgKfOS01VklT0tKeqTL4LiBLdLKO0a6+x9nauezsOdeiN","nonce":"zhiRucguKoeZLdX2K51x/Ywdy5iQlGwu"}
keepass.js:1308 Error: No content script available for this tab.
keepass.js:120 Got message: {"action":"get-databasehash","message":"aWW0CptfUvTqlsHfYBtvYekPJIhouF4blS2mUNnxN/Go5F8W6cOrYgGSMY/FRYzSH25dWyAVgkRUD9n/3r8pemhERPX/12px4L0GmuToOGUUkekGW0VKkQTnxq3mecxoGNKljQofgIKzDojnSJVVItext5lcHUvyJwm0rw/uCgMMnHFluqHTJOUFDGp0aPgfKBLOGnIhz5doMs+mM0Nzj8EMG+ia71Pfd8yQop8/ngw1DKWfKBkczWdIv9ir/k0ugpEEcJAbtA/pwoxfSVEQ23fwLvAvRPzImK0K7iCTUnzbv4ZfqMoeoFTbVpHT+FEhM0RNRiXvsb+CwDdrghlI//lbZoYcxyTtGMvl+n0teSzp/RHQt41UIWRTh3me1wY=","nonce":"1mWQDUO1c6qMIgReiU6h8DHxzXiZYEdH"}
keepass.js:120 Got message: {"action":"get-logins","message":"yB22SwqMnxvfsIVkbPqSonEnaYLK0TORYXRtc79IRITecewAaUWny69bWX6cpht5cEz+cH0YX7vBzktVsmLQuj+lGNx/0VER39e4K8ezUHMjTFz4sCtMM7bJewyB1TJ/QAsgLuYibTJGt4t/9AAASq9CXc7z/F0Ae6QRzdYyKaP1DEjLjCzp6eDGDCO2EckECBWaNkictPf7Z+LGrrSV+UI1v5PcXAC5+2eJHdD8Z9uNCCtPbXudX3lMVAy7OIyNjJ9CQ8KDvDLwva1GH906eoK2BRO3rr03+tLs3uBrf4jnviqgzoJ3P7exfti8QfSy38i/55lAsA4BzKFDGCLiTyM1nR3XbcfHTEeR+zZb9wVB47Z5vJCUJn0xYwfq4dNTLpo6pd8ujmISZx30EbJOgPCyS12fW6btXcQ09RdnEidgsN2DVL2ootL84/VIXI0AyWO3AZeHUdV8FoznFVv2U0RLoD1r2Qe/KVfVsfW29l3TfmD09jvLSsx0FnBb0lFKgSVBdumeBE3ZQMSb/VvSlq/TyCVj5ArcbQ9MkCk5QK2VxTaMQmTvC0qgyZ4+mbm3eBoh4yiCs2FcgIGNS1xUy6zpKz7AQioUnmc3y/1hBIKrwS+ZifsN267Uhc/hjDogqpBVMX2pjxRL5niPym7DYXdoxlONw4KBX9d73fhvYMc4NGz0rPYORnvjHlXbyPkP1Of1CGBWymx+anu/7ujfgEhrfSAMw3V51qdXPtlEycmLQVWu2Cm63KrpsbEjkA/qyJH1kARvm165gPwpPY0Ofoaceds9ePge0/C8rX3/No0+1xHVjOEtVBxNB9+fTHxxV4Ed42w0PVglHfMgPBxaYG6hwGNTgVqhXV7bfQfa4E8wEAqAXNJx1POE/jKLyCYKU6RiZG0O8DDd8nvbJrHggFaAEYaCWz5Oeuq0zG0/MRYn9XXyyBF2/5dMVp7SKnYJqNM5QWy2iOd4pzvZWpQi+zooJIOlE0fsLsxUeWqy9DJtc7A4gxF5g+syaYE2cqiP8i7vQJLHHwprsGk0G5+Z1ymjQN7TC8kd1fKtKtbXQLC/3tnKYrU7wzwTSRLQ4Ce2tkE+oDvauFYbXg7BZSe9jAW+wglYPl+6CYyVZjpmxGrQvdS3bAOlbL/ReAIHP/Sq23TMk87aYtw3W40CD/I=","nonce":"eYj7019WvYX7F5Ol7sPjgl2CqcSPmNjH"}
3keepass.js:120 Got message: {"action":"get-databasehash","message":"3W6ULtO9ipjqHsfQXFqNYwdUCxlPmnMZsU0WFgNuHF+1uNY8d9GPX/R+8ErFrW3GZk6G4ptPIxPqYQS59VyGhGCbUnf+9vNbXQ83+I4uZW0IOptA0iYGTqixaEuit9jk6F4OEfuNoEcXbQpxDufrfX/9aPNoTbltsd2VAgA+8AyXmruO+T2vjQioduiNrr7ODsu50qnj4DrLmrYlFf4cnKCZ0X057T/XpUz1Tx554IHtGnwsisbzeMly/2KDjRfQbMkoZvtZaH3z+79ZWn5rI4r/JEe9pwzp8oplcKmCt/IyVH/DXDLMX1y83Gc2pVFH9JWaXGjdgyjz0X2bj7ozbcYDyd3IDLBVMRmfDZQZxDvtSvASw8QXD33MZIDBS3M=","nonce":"oIut2KukCn9gGMH4v+e4kNL5JTD9zfo8"}
3keepass.js:120 Got message: {"action":"get-databasehash","message":"jF6czq/dgNY/KX1pfapoTOFzdD2yHtAWnx/hTUi6jxJ4XmkDxYTu7HHmSi2chLb1URIet5AFGeqfTKIdR2JD9+AvGACIStOGnJm7cHWxvhYQM6cGOmf+qDULFwAhB1bjl/nwc2bshZa1ZTcqYBlWyB04qLonbDX98yY3ZwMrwa2GQQjJO2KsWGP6cuSeqNGiewMlQb78eCcGfDDMUD8lTsOD6CbJQj224cCjA/qCGULAqFQuF1B83jGrJIPOugxyAtlDUCwmZzAM5dqstk5wLy9BlC5OEdoW6ofOpyfmeTnza+bkEQ7ERifuQdBT16uDx7P2agpBUeTp8Sgy1fn5wVSzmJ9TA27dGQnXWhrX29ClClotLmTam44O1ylDvhY=","nonce":"rT5EoFM5cleIAhwXZWe+TFrJpHHjwIav"}
3keepass.js:120 Got message: {"action":"get-databasehash","message":"7j2lxVmimni4zp9vUrT0IWc1Nvt4kPreuTtCBmXN1SfSg5nvKAaTUDgm7BQJhYvYl8374JgW0h+6wSZjPsXGkuG+ydXXkqLgQieNuyhptMswu/FfhcqJ51mrwllFhrlsP9tAeAEZhAFTYYJsY3dN0rS9xIZ0MApu53IyhzpIxA99u74Y73L1ULPPp1G+76MnYMv3+yIO78d63c1kFiCkARDHut2wDPvPclpz4zpW38EuUY9qui889gMmkQvNaCvhdgayD1GMcoTG8na6VNwQ+xqqGj+NWQrMxrra051/1jrz8wpAHP56Wq6Jbj2kJzQGgOf/h9o30/s4PVvcksHthZaGvn+0H1GIQu4itbd7enB1J6msJ/p+qS9Tib6GYLA=","nonce":"fshSU5bgPlr4FfUzf3CTAW9dsmQjiNd3"}
3keepass.js:120 Got message: {}
3keepass.js:120 Got message: {"action":"generate-password","message":"Z+sN+w1MzihFvwB6uFiZK9tjY68fuDVkNlrKMg8CsR5d1qkNXinQZT6UFr3t34PY9ISRkw0I0pTdK16dTf40tBRdH/BBj4Olz4f67awdM8Aw/sj2VXGw9RKaK4GRnFf3D7PFZCj7kxlVC5p2K16Yoo/ohOpaHWU/QKXtN2vnsHxU9Xm/H1XHcNzVXVTgntOcCHvVpspcYbkS1PZ2EtLLufQWdfc=","nonce":"1sAjeFu//GwrC4WCR9qJM7nzx5OWxsPs"}
keepass.js:362 Retrieved password:  {IR5:{zei<t'WEaX<Nh/
5keepass.js:120 Got message: {"action":"get-databasehash","message":"y3hnYDKM4UCuv60+13b4xv8l1H1QWITyREXShqJv51CNJ25Xri7F6Se2JfOCO/F/KldZIE4cI+2gUauKP9Pc5gt/kbHLeQnjFmQNi/fH582IHAgJQ6a/8yJvqZjgHfZeu2dKX8AbE3QnHk+/LUlniRAp1fkwo0Y3ophoAaKfG8Z8Wae0cGSRW9VlheoTHf9i1nfY27aEjDp11JszjeOZ9Sqpl9exgd01s9uQSYDkQ9Y3xcfTwh29g8YW0CvxVOO5WVzXk6yNlTIc3H2quqzbgUrAZHZvWqn1fBSJuRyV/X8hLPefchrGvnvwI3iwbCLwXIxzItR3roYzZzseicCvO+oAERxlAppHLbmuj46auDReeSv8n41BE+9P1TJ82QI=","nonce":"+8a68fazaht2vMwF+sMIQ2LZu6N5XP4U"}
keepass.js:1172 Error: Nonce compare failed
5keepass.js:120 Got message: {"action":"get-databasehash","message":"MX/j8z7QP0Hgm3LBWCxJm1/P76c1QFDnuE87OUVs58nuUz+e71z+elBj18fl34vMIxHzRZ/rB870lWQfFUGAXibL8hPYAuzfz3qmwVm++xohj/qvBjmGbOvh7QIok5pkPSYvNJpNOgHBu7qvzfwaggo27kJzoYKSbPjoFg5n/AeI1r4wWh1okeWFxdpzZ5ckYGh+8hsL3EvW9BcdJ1KCBg8Hcsp3Y3S0FNIQtZmlYc5QE8YXafiT6dLDEdCM2g/mi7+lKOyBamJkYt0G6jIy/NMp+f1QFigPyebeGINIwSep4RuB3MtzPdzMOlDAsyr9U3/czTK/+7g5gvRsMC2FMGp4W2BXwqH01b43lV2GPCs18I0/s2niodDGi5Xnp6I=","nonce":"VOBRmi3yvTatGkg8x9o8JfRJxUlmABL8"}
keepass.js:1172 Error: Nonce compare failed
5keepass.js:120 Got message: {}
5keepass.js:120 Got message: {"action":"generate-password","message":"txv/NC11kYrnbFVITnEs+RWvroVvK4MLmnHKaggqzjkMkQYZSz6tw7chAh4GdzIujJcKK18r+PyZhTdSvTutamfKkYgndf3fMDFL0a5luaA4b96mrK+UijOF9Ym5Iz4UoIGxAYlDJgkV5/my80UTXl6cKCcMUJIMTryKL0AcmSBetp02PNrlS9srR/VU1cy8HiTHUB8O3l2RnOcYFgRxTliNfR8h","nonce":"7dLzMvV650FwzQavP28eUOzpWtwnQ6Wp"}
keepass.js:1148 Error: Nonce compare failed
keepass.js:360 GeneratePassword rejected
```

--- a/src/browser/BrowserAction.cpp
+++ b/src/browser/BrowserAction.cpp
@@ -292,9 +292,18 @@ QJsonObject BrowserAction::handleGeneratePassword(QLocalSocket* socket, const QJ
         return getErrorReply(action, ERROR_KEEPASS_CANNOT_DECRYPT_MESSAGE);
     }
 
-    // Do not allow multiple requests from the same client. Response with an empty password.
+    auto requestId = decrypted.value("requestID").toString();
+
+    // Do not allow multiple requests from the same client
     if (browserService()->isPasswordGeneratorRequested()) {
-        return getErrorReply(action, ERROR_KEEPASS_ACTION_CANCELLED_OR_DENIED);
+        auto errorReply = getErrorReply(action, ERROR_KEEPASS_ACTION_CANCELLED_OR_DENIED);
+
+        // Append requestID to the response if found
+        if (!requestId.isEmpty()) {
+            errorReply["requestID"] = requestId;
+        }
+
+        return errorReply;
     }
 
     browserService()->showPasswordGenerator(socket, incrementedNonce, m_clientPublicKey, m_secretKey);

--- a/src/browser/BrowserAction.cpp
+++ b/src/browser/BrowserAction.cpp
@@ -27,10 +27,11 @@
 #include <QJsonArray>
 #include <QJsonDocument>
 #include <QJsonObject>
+#include <QLocalSocket>
 
 const int BrowserAction::MaxUrlLength = 256;
 
-QJsonObject BrowserAction::processClientMessage(const QJsonObject& json)
+QJsonObject BrowserAction::processClientMessage(QLocalSocket* socket, const QJsonObject& json)
 {
     if (json.isEmpty()) {
         return getErrorReply("", ERROR_KEEPASS_EMPTY_MESSAGE_RECEIVED);
@@ -56,13 +57,13 @@ QJsonObject BrowserAction::processClientMessage(const QJsonObject& json)
         }
     }
 
-    return handleAction(json);
+    return handleAction(socket, json);
 }
 
 // Private functions
 ///////////////////////
 
-QJsonObject BrowserAction::handleAction(const QJsonObject& json)
+QJsonObject BrowserAction::handleAction(QLocalSocket* socket, const QJsonObject& json)
 {
     QString action = json.value("action").toString();
 
@@ -77,7 +78,7 @@ QJsonObject BrowserAction::handleAction(const QJsonObject& json)
     } else if (action.compare("get-logins") == 0) {
         return handleGetLogins(json, action);
     } else if (action.compare("generate-password") == 0) {
-        return handleGeneratePassword(json, action);
+        return handleGeneratePassword(socket, json, action);
     } else if (action.compare("set-login") == 0) {
         return handleSetLogin(json, action);
     } else if (action.compare("lock-database") == 0) {
@@ -279,20 +280,24 @@ QJsonObject BrowserAction::handleGetLogins(const QJsonObject& json, const QStrin
     return buildResponse(action, message, newNonce);
 }
 
-QJsonObject BrowserAction::handleGeneratePassword(const QJsonObject& json, const QString& action)
+QJsonObject BrowserAction::handleGeneratePassword(QLocalSocket* socket, const QJsonObject& json, const QString& action)
 {
     auto errorMessage = getErrorReply(action, ERROR_KEEPASS_ACTION_CANCELLED_OR_DENIED);
     auto nonce = json.value("nonce").toString();
     auto incrementedNonce = browserMessageBuilder()->incrementNonce(nonce);
 
-    // Do not allow multiple requests from the same client. Response with an empty password.
-    if (browserService()->isPasswordGeneratorRequested()) {
-        QJsonObject message = browserMessageBuilder()->buildMessage(incrementedNonce);
-        message["password"] = "";
-        return buildResponse("generate-password", message, incrementedNonce);
+    const QString encrypted = json.value("message").toString();
+    const QJsonObject decrypted = decryptMessage(encrypted, nonce);
+    if (decrypted.isEmpty()) {
+        return getErrorReply(action, ERROR_KEEPASS_CANNOT_DECRYPT_MESSAGE);
     }
 
-    browserService()->showPasswordGenerator(incrementedNonce, m_clientPublicKey, m_secretKey);
+    // Do not allow multiple requests from the same client. Response with an empty password.
+    if (browserService()->isPasswordGeneratorRequested()) {
+        return getErrorReply(action, ERROR_KEEPASS_ACTION_CANCELLED_OR_DENIED);
+    }
+
+    browserService()->showPasswordGenerator(socket, incrementedNonce, m_clientPublicKey, m_secretKey);
     return QJsonObject();
 }
 

--- a/src/browser/BrowserAction.cpp
+++ b/src/browser/BrowserAction.cpp
@@ -1,5 +1,5 @@
 /*
- *  Copyright (C) 2021 KeePassXC Team <team@keepassxc.org>
+ *  Copyright (C) 2022 KeePassXC Team <team@keepassxc.org>
  *
  *  This program is free software: you can redistribute it and/or modify
  *  it under the terms of the GNU General Public License as published by
@@ -16,63 +16,19 @@
  */
 
 #include "BrowserAction.h"
-
+#include "BrowserMessageBuilder.h"
 #include "BrowserService.h"
 #include "BrowserSettings.h"
 #include "BrowserShared.h"
 #include "config-keepassx.h"
 #include "core/Global.h"
 #include "core/Tools.h"
-#include "gui/PasswordGeneratorWidget.h"
 
 #include <QJsonArray>
 #include <QJsonDocument>
 #include <QJsonObject>
 
-#include <botan/sodium.h>
-
-using namespace Botan::Sodium;
-
-namespace
-{
-    enum
-    {
-        ERROR_KEEPASS_DATABASE_NOT_OPENED = 1,
-        ERROR_KEEPASS_DATABASE_HASH_NOT_RECEIVED = 2,
-        ERROR_KEEPASS_CLIENT_PUBLIC_KEY_NOT_RECEIVED = 3,
-        ERROR_KEEPASS_CANNOT_DECRYPT_MESSAGE = 4,
-        ERROR_KEEPASS_TIMEOUT_OR_NOT_CONNECTED = 5,
-        ERROR_KEEPASS_ACTION_CANCELLED_OR_DENIED = 6,
-        ERROR_KEEPASS_CANNOT_ENCRYPT_MESSAGE = 7,
-        ERROR_KEEPASS_ASSOCIATION_FAILED = 8,
-        ERROR_KEEPASS_KEY_CHANGE_FAILED = 9,
-        ERROR_KEEPASS_ENCRYPTION_KEY_UNRECOGNIZED = 10,
-        ERROR_KEEPASS_NO_SAVED_DATABASES_FOUND = 11,
-        ERROR_KEEPASS_INCORRECT_ACTION = 12,
-        ERROR_KEEPASS_EMPTY_MESSAGE_RECEIVED = 13,
-        ERROR_KEEPASS_NO_URL_PROVIDED = 14,
-        ERROR_KEEPASS_NO_LOGINS_FOUND = 15,
-        ERROR_KEEPASS_NO_GROUPS_FOUND = 16,
-        ERROR_KEEPASS_CANNOT_CREATE_NEW_GROUP = 17,
-        ERROR_KEEPASS_NO_VALID_UUID_PROVIDED = 18
-    };
-}
-
 const int BrowserAction::MaxUrlLength = 256;
-
-BrowserAction::BrowserAction()
-{
-    QObject::connect(browserService(),
-                     &BrowserService::passwordGenerated,
-                     browserService(),
-                     [=](const QString& password, const QString& nonce) {
-                         auto newNonce = incrementNonce(nonce);
-                         QJsonObject message = buildMessage(newNonce);
-                         message["password"] = password;
-
-                         browserService()->sendPassword(buildResponse("generate-password", message, newNonce));
-                     });
-}
 
 QJsonObject BrowserAction::processClientMessage(const QJsonObject& json)
 {
@@ -152,23 +108,18 @@ QJsonObject BrowserAction::handleChangePublicKeys(const QJsonObject& json, const
     }
 
     m_associated = false;
-    unsigned char pk[crypto_box_PUBLICKEYBYTES];
-    unsigned char sk[crypto_box_SECRETKEYBYTES];
-    crypto_box_keypair(pk, sk);
-
-    const QString publicKey = getBase64FromKey(pk, crypto_box_PUBLICKEYBYTES);
-    const QString secretKey = getBase64FromKey(sk, crypto_box_SECRETKEYBYTES);
-    if (publicKey.isEmpty() || secretKey.isEmpty()) {
+    auto keyPair = browserMessageBuilder()->getKeyPair();
+    if (keyPair.first.isEmpty() || keyPair.second.isEmpty()) {
         return getErrorReply(action, ERROR_KEEPASS_ENCRYPTION_KEY_UNRECOGNIZED);
     }
 
     m_clientPublicKey = clientPublicKey;
-    m_publicKey = publicKey;
-    m_secretKey = secretKey;
+    m_publicKey = keyPair.first;
+    m_secretKey = keyPair.second;
 
-    QJsonObject response = buildMessage(incrementNonce(nonce));
+    QJsonObject response = browserMessageBuilder()->buildMessage(browserMessageBuilder()->incrementNonce(nonce));
     response["action"] = action;
-    response["publicKey"] = publicKey;
+    response["publicKey"] = keyPair.first;
 
     return response;
 }
@@ -190,9 +141,9 @@ QJsonObject BrowserAction::handleGetDatabaseHash(const QJsonObject& json, const 
 
     QString command = decrypted.value("action").toString();
     if (!command.isEmpty() && command.compare("get-databasehash") == 0) {
-        const QString newNonce = incrementNonce(nonce);
+        const QString newNonce = browserMessageBuilder()->incrementNonce(nonce);
 
-        QJsonObject message = buildMessage(newNonce);
+        QJsonObject message = browserMessageBuilder()->buildMessage(newNonce);
         message["hash"] = hash;
 
         // Update a legacy database hash if found
@@ -236,9 +187,9 @@ QJsonObject BrowserAction::handleAssociate(const QJsonObject& json, const QStrin
         }
 
         m_associated = true;
-        const QString newNonce = incrementNonce(nonce);
+        const QString newNonce = browserMessageBuilder()->incrementNonce(nonce);
 
-        QJsonObject message = buildMessage(newNonce);
+        QJsonObject message = browserMessageBuilder()->buildMessage(newNonce);
         message["hash"] = hash;
         message["id"] = id;
         return buildResponse(action, message, newNonce);
@@ -270,9 +221,9 @@ QJsonObject BrowserAction::handleTestAssociate(const QJsonObject& json, const QS
     }
 
     m_associated = true;
-    const QString newNonce = incrementNonce(nonce);
+    const QString newNonce = browserMessageBuilder()->incrementNonce(nonce);
 
-    QJsonObject message = buildMessage(newNonce);
+    QJsonObject message = browserMessageBuilder()->buildMessage(newNonce);
     message["hash"] = hash;
     message["id"] = id;
 
@@ -317,9 +268,9 @@ QJsonObject BrowserAction::handleGetLogins(const QJsonObject& json, const QStrin
         return getErrorReply(action, ERROR_KEEPASS_NO_LOGINS_FOUND);
     }
 
-    const QString newNonce = incrementNonce(nonce);
+    const QString newNonce = browserMessageBuilder()->incrementNonce(nonce);
 
-    QJsonObject message = buildMessage(newNonce);
+    QJsonObject message = browserMessageBuilder()->buildMessage(newNonce);
     message["count"] = users.count();
     message["entries"] = users;
     message["hash"] = hash;
@@ -332,8 +283,16 @@ QJsonObject BrowserAction::handleGeneratePassword(const QJsonObject& json, const
 {
     auto errorMessage = getErrorReply(action, ERROR_KEEPASS_ACTION_CANCELLED_OR_DENIED);
     auto nonce = json.value("nonce").toString();
+    auto incrementedNonce = browserMessageBuilder()->incrementNonce(nonce);
 
-    browserService()->showPasswordGenerator(errorMessage, nonce);
+    // Do not allow multiple requests from the same client. Response with an empty password.
+    if (browserService()->isPasswordGeneratorRequested()) {
+        QJsonObject message = browserMessageBuilder()->buildMessage(incrementedNonce);
+        message["password"] = "";
+        return buildResponse("generate-password", message, incrementedNonce);
+    }
+
+    browserService()->showPasswordGenerator(incrementedNonce, m_clientPublicKey, m_secretKey);
     return QJsonObject();
 }
 
@@ -379,9 +338,9 @@ QJsonObject BrowserAction::handleSetLogin(const QJsonObject& json, const QString
         result = browserService()->updateEntry(id, uuid, login, password, url, submitUrl);
     }
 
-    const QString newNonce = incrementNonce(nonce);
+    const QString newNonce = browserMessageBuilder()->incrementNonce(nonce);
 
-    QJsonObject message = buildMessage(newNonce);
+    QJsonObject message = browserMessageBuilder()->buildMessage(newNonce);
     message["count"] = QJsonValue::Null;
     message["entries"] = QJsonValue::Null;
     message["error"] = result ? QStringLiteral("success") : QStringLiteral("error");
@@ -409,8 +368,8 @@ QJsonObject BrowserAction::handleLockDatabase(const QJsonObject& json, const QSt
     if (!command.isEmpty() && command.compare("lock-database") == 0) {
         browserService()->lockDatabase();
 
-        const QString newNonce = incrementNonce(nonce);
-        QJsonObject message = buildMessage(newNonce);
+        const QString newNonce = browserMessageBuilder()->incrementNonce(nonce);
+        QJsonObject message = browserMessageBuilder()->buildMessage(newNonce);
 
         return buildResponse(action, message, newNonce);
     }
@@ -443,9 +402,9 @@ QJsonObject BrowserAction::handleGetDatabaseGroups(const QJsonObject& json, cons
         return getErrorReply(action, ERROR_KEEPASS_NO_GROUPS_FOUND);
     }
 
-    const QString newNonce = incrementNonce(nonce);
+    const QString newNonce = browserMessageBuilder()->incrementNonce(nonce);
 
-    QJsonObject message = buildMessage(newNonce);
+    QJsonObject message = browserMessageBuilder()->buildMessage(newNonce);
     message["groups"] = groups;
 
     return buildResponse(action, message, newNonce);
@@ -477,9 +436,9 @@ QJsonObject BrowserAction::handleCreateNewGroup(const QJsonObject& json, const Q
         return getErrorReply(action, ERROR_KEEPASS_CANNOT_CREATE_NEW_GROUP);
     }
 
-    const QString newNonce = incrementNonce(nonce);
+    const QString newNonce = browserMessageBuilder()->incrementNonce(nonce);
 
-    QJsonObject message = buildMessage(newNonce);
+    QJsonObject message = browserMessageBuilder()->buildMessage(newNonce);
     message["name"] = newGroup["name"];
     message["uuid"] = newGroup["uuid"];
 
@@ -512,9 +471,9 @@ QJsonObject BrowserAction::handleGetTotp(const QJsonObject& json, const QString&
 
     // Get the current TOTP
     const auto totp = browserService()->getCurrentTotp(uuid);
-    const QString newNonce = incrementNonce(nonce);
+    const QString newNonce = browserMessageBuilder()->incrementNonce(nonce);
 
-    QJsonObject message = buildMessage(newNonce);
+    QJsonObject message = browserMessageBuilder()->buildMessage(newNonce);
     message["totp"] = totp;
 
     return buildResponse(action, message, newNonce);
@@ -546,8 +505,8 @@ QJsonObject BrowserAction::handleDeleteEntry(const QJsonObject& json, const QStr
 
     const auto result = browserService()->deleteEntry(uuid);
 
-    const QString newNonce = incrementNonce(nonce);
-    QJsonObject message = buildMessage(newNonce);
+    const QString newNonce = browserMessageBuilder()->incrementNonce(nonce);
+    QJsonObject message = browserMessageBuilder()->buildMessage(newNonce);
     message["success"] = result ? TRUE_STR : FALSE_STR;
 
     return buildResponse(action, message, newNonce);
@@ -575,202 +534,22 @@ QJsonObject BrowserAction::handleGlobalAutoType(const QJsonObject& json, const Q
 
     browserService()->requestGlobalAutoType(topLevelDomain);
 
-    const QString newNonce = incrementNonce(nonce);
-    QJsonObject message = buildMessage(newNonce);
+    const QString newNonce = browserMessageBuilder()->incrementNonce(nonce);
+    QJsonObject message = browserMessageBuilder()->buildMessage(newNonce);
     return buildResponse(action, message, newNonce);
-}
-
-QJsonObject BrowserAction::getErrorReply(const QString& action, const int errorCode) const
-{
-    QJsonObject response;
-    response["action"] = action;
-    response["errorCode"] = QString::number(errorCode);
-    response["error"] = getErrorMessage(errorCode);
-    return response;
-}
-
-QJsonObject BrowserAction::buildMessage(const QString& nonce) const
-{
-    QJsonObject message;
-    message["version"] = KEEPASSXC_VERSION;
-    message["success"] = TRUE_STR;
-    message["nonce"] = nonce;
-    return message;
-}
-
-QJsonObject BrowserAction::buildResponse(const QString& action, const QJsonObject& message, const QString& nonce)
-{
-    QJsonObject response;
-    QString encryptedMessage = encryptMessage(message, nonce);
-    if (encryptedMessage.isEmpty()) {
-        return getErrorReply(action, ERROR_KEEPASS_CANNOT_ENCRYPT_MESSAGE);
-    }
-
-    response["action"] = action;
-    response["message"] = encryptedMessage;
-    response["nonce"] = nonce;
-    return response;
-}
-
-QString BrowserAction::getErrorMessage(const int errorCode) const
-{
-    switch (errorCode) {
-    case ERROR_KEEPASS_DATABASE_NOT_OPENED:
-        return QObject::tr("Database not opened");
-    case ERROR_KEEPASS_DATABASE_HASH_NOT_RECEIVED:
-        return QObject::tr("Database hash not available");
-    case ERROR_KEEPASS_CLIENT_PUBLIC_KEY_NOT_RECEIVED:
-        return QObject::tr("Client public key not received");
-    case ERROR_KEEPASS_CANNOT_DECRYPT_MESSAGE:
-        return QObject::tr("Cannot decrypt message");
-    case ERROR_KEEPASS_ACTION_CANCELLED_OR_DENIED:
-        return QObject::tr("Action cancelled or denied");
-    case ERROR_KEEPASS_CANNOT_ENCRYPT_MESSAGE:
-        return QObject::tr("Message encryption failed.");
-    case ERROR_KEEPASS_ASSOCIATION_FAILED:
-        return QObject::tr("KeePassXC association failed, try again");
-    case ERROR_KEEPASS_ENCRYPTION_KEY_UNRECOGNIZED:
-        return QObject::tr("Encryption key is not recognized");
-    case ERROR_KEEPASS_INCORRECT_ACTION:
-        return QObject::tr("Incorrect action");
-    case ERROR_KEEPASS_EMPTY_MESSAGE_RECEIVED:
-        return QObject::tr("Empty message received");
-    case ERROR_KEEPASS_NO_URL_PROVIDED:
-        return QObject::tr("No URL provided");
-    case ERROR_KEEPASS_NO_LOGINS_FOUND:
-        return QObject::tr("No logins found");
-    case ERROR_KEEPASS_NO_GROUPS_FOUND:
-        return QObject::tr("No groups found");
-    case ERROR_KEEPASS_CANNOT_CREATE_NEW_GROUP:
-        return QObject::tr("Cannot create new group");
-    case ERROR_KEEPASS_NO_VALID_UUID_PROVIDED:
-        return QObject::tr("No valid UUID provided");
-    default:
-        return QObject::tr("Unknown error");
-    }
-}
-
-QString BrowserAction::encryptMessage(const QJsonObject& message, const QString& nonce)
-{
-    if (message.isEmpty() || nonce.isEmpty()) {
-        return QString();
-    }
-
-    const QString reply(QJsonDocument(message).toJson());
-    if (!reply.isEmpty()) {
-        return encrypt(reply, nonce);
-    }
-
-    return QString();
 }
 
 QJsonObject BrowserAction::decryptMessage(const QString& message, const QString& nonce)
 {
-    if (message.isEmpty() || nonce.isEmpty()) {
-        return QJsonObject();
-    }
-
-    QByteArray ba = decrypt(message, nonce);
-    if (ba.isEmpty()) {
-        return QJsonObject();
-    }
-
-    return getJsonObject(ba);
+    return browserMessageBuilder()->decryptMessage(message, nonce, m_clientPublicKey, m_secretKey);
 }
 
-QString BrowserAction::encrypt(const QString& plaintext, const QString& nonce)
+QJsonObject BrowserAction::getErrorReply(const QString& action, const int errorCode) const
 {
-    const QByteArray ma = plaintext.toUtf8();
-    const QByteArray na = base64Decode(nonce);
-    const QByteArray ca = base64Decode(m_clientPublicKey);
-    const QByteArray sa = base64Decode(m_secretKey);
-
-    std::vector<unsigned char> m(ma.cbegin(), ma.cend());
-    std::vector<unsigned char> n(na.cbegin(), na.cend());
-    std::vector<unsigned char> ck(ca.cbegin(), ca.cend());
-    std::vector<unsigned char> sk(sa.cbegin(), sa.cend());
-
-    std::vector<unsigned char> e;
-    e.resize(BrowserShared::NATIVEMSG_MAX_LENGTH);
-
-    if (m.empty() || n.empty() || ck.empty() || sk.empty()) {
-        return QString();
-    }
-
-    if (crypto_box_easy(e.data(), m.data(), m.size(), n.data(), ck.data(), sk.data()) == 0) {
-        QByteArray res = getQByteArray(e.data(), (crypto_box_MACBYTES + ma.length()));
-        return res.toBase64();
-    }
-
-    return QString();
+    return browserMessageBuilder()->getErrorReply(action, errorCode);
 }
 
-QByteArray BrowserAction::decrypt(const QString& encrypted, const QString& nonce)
+QJsonObject BrowserAction::buildResponse(const QString& action, const QJsonObject& message, const QString& nonce)
 {
-    const QByteArray ma = base64Decode(encrypted);
-    const QByteArray na = base64Decode(nonce);
-    const QByteArray ca = base64Decode(m_clientPublicKey);
-    const QByteArray sa = base64Decode(m_secretKey);
-
-    std::vector<unsigned char> m(ma.cbegin(), ma.cend());
-    std::vector<unsigned char> n(na.cbegin(), na.cend());
-    std::vector<unsigned char> ck(ca.cbegin(), ca.cend());
-    std::vector<unsigned char> sk(sa.cbegin(), sa.cend());
-
-    std::vector<unsigned char> d;
-    d.resize(BrowserShared::NATIVEMSG_MAX_LENGTH);
-
-    if (m.empty() || n.empty() || ck.empty() || sk.empty()) {
-        return QByteArray();
-    }
-
-    if (crypto_box_open_easy(d.data(), m.data(), ma.length(), n.data(), ck.data(), sk.data()) == 0) {
-        return getQByteArray(d.data(), std::char_traits<char>::length(reinterpret_cast<const char*>(d.data())));
-    }
-
-    return QByteArray();
-}
-
-QString BrowserAction::getBase64FromKey(const uchar* array, const uint len)
-{
-    return getQByteArray(array, len).toBase64();
-}
-
-QByteArray BrowserAction::getQByteArray(const uchar* array, const uint len) const
-{
-    QByteArray qba;
-    qba.reserve(len);
-    for (uint i = 0; i < len; ++i) {
-        qba.append(static_cast<char>(array[i]));
-    }
-    return qba;
-}
-
-QJsonObject BrowserAction::getJsonObject(const uchar* pArray, const uint len) const
-{
-    QByteArray arr = getQByteArray(pArray, len);
-    QJsonParseError err;
-    QJsonDocument doc(QJsonDocument::fromJson(arr, &err));
-    return doc.object();
-}
-
-QJsonObject BrowserAction::getJsonObject(const QByteArray& ba) const
-{
-    QJsonParseError err;
-    QJsonDocument doc(QJsonDocument::fromJson(ba, &err));
-    return doc.object();
-}
-
-QByteArray BrowserAction::base64Decode(const QString& str)
-{
-    return QByteArray::fromBase64(str.toUtf8());
-}
-
-QString BrowserAction::incrementNonce(const QString& nonce)
-{
-    const QByteArray nonceArray = base64Decode(nonce);
-    std::vector<unsigned char> n(nonceArray.cbegin(), nonceArray.cend());
-
-    sodium_increment(n.data(), n.size());
-    return getQByteArray(n.data(), n.size()).toBase64();
+    return browserMessageBuilder()->buildResponse(action, message, nonce, m_clientPublicKey, m_secretKey);
 }

--- a/src/browser/BrowserAction.h
+++ b/src/browser/BrowserAction.h
@@ -1,5 +1,5 @@
 /*
- *  Copyright (C) 2021 KeePassXC Team <team@keepassxc.org>
+ *  Copyright (C) 2022 KeePassXC Team <team@keepassxc.org>
  *
  *  This program is free software: you can redistribute it and/or modify
  *  it under the terms of the GNU General Public License as published by
@@ -25,7 +25,7 @@ class QJsonObject;
 class BrowserAction
 {
 public:
-    explicit BrowserAction();
+    explicit BrowserAction() = default;
     ~BrowserAction() = default;
 
     QJsonObject processClientMessage(const QJsonObject& json);
@@ -46,22 +46,11 @@ private:
     QJsonObject handleDeleteEntry(const QJsonObject& json, const QString& action);
     QJsonObject handleGlobalAutoType(const QJsonObject& json, const QString& action);
 
+private:
     QJsonObject buildMessage(const QString& nonce) const;
     QJsonObject buildResponse(const QString& action, const QJsonObject& message, const QString& nonce);
     QJsonObject getErrorReply(const QString& action, const int errorCode) const;
-    QString getErrorMessage(const int errorCode) const;
-
-    QString encryptMessage(const QJsonObject& message, const QString& nonce);
     QJsonObject decryptMessage(const QString& message, const QString& nonce);
-    QString encrypt(const QString& plaintext, const QString& nonce);
-    QByteArray decrypt(const QString& encrypted, const QString& nonce);
-
-    QString getBase64FromKey(const uchar* array, const uint len);
-    QByteArray getQByteArray(const uchar* array, const uint len) const;
-    QJsonObject getJsonObject(const uchar* pArray, const uint len) const;
-    QJsonObject getJsonObject(const QByteArray& ba) const;
-    QByteArray base64Decode(const QString& str);
-    QString incrementNonce(const QString& nonce);
 
 private:
     static const int MaxUrlLength;

--- a/src/browser/BrowserAction.h
+++ b/src/browser/BrowserAction.h
@@ -21,6 +21,7 @@
 #include <QString>
 
 class QJsonObject;
+class QLocalSocket;
 
 class BrowserAction
 {
@@ -28,16 +29,16 @@ public:
     explicit BrowserAction() = default;
     ~BrowserAction() = default;
 
-    QJsonObject processClientMessage(const QJsonObject& json);
+    QJsonObject processClientMessage(QLocalSocket* socket, const QJsonObject& json);
 
 private:
-    QJsonObject handleAction(const QJsonObject& json);
+    QJsonObject handleAction(QLocalSocket* socket, const QJsonObject& json);
     QJsonObject handleChangePublicKeys(const QJsonObject& json, const QString& action);
     QJsonObject handleGetDatabaseHash(const QJsonObject& json, const QString& action);
     QJsonObject handleAssociate(const QJsonObject& json, const QString& action);
     QJsonObject handleTestAssociate(const QJsonObject& json, const QString& action);
     QJsonObject handleGetLogins(const QJsonObject& json, const QString& action);
-    QJsonObject handleGeneratePassword(const QJsonObject& json, const QString& action);
+    QJsonObject handleGeneratePassword(QLocalSocket* socket, const QJsonObject& json, const QString& action);
     QJsonObject handleSetLogin(const QJsonObject& json, const QString& action);
     QJsonObject handleLockDatabase(const QJsonObject& json, const QString& action);
     QJsonObject handleGetDatabaseGroups(const QJsonObject& json, const QString& action);

--- a/src/browser/BrowserHost.cpp
+++ b/src/browser/BrowserHost.cpp
@@ -88,18 +88,29 @@ void BrowserHost::readProxyMessage()
         return;
     }
 
-    emit clientMessageReceived(json.object());
+    emit clientMessageReceived(socket, json.object());
 }
 
-void BrowserHost::sendClientMessage(const QJsonObject& json)
+void BrowserHost::broadcastClientMessage(const QJsonObject& json)
 {
     QString reply(QJsonDocument(json).toJson(QJsonDocument::Compact));
     for (const auto socket : m_socketList) {
-        if (socket && socket->isValid() && socket->state() == QLocalSocket::ConnectedState) {
-            QByteArray arr = reply.toUtf8();
-            socket->write(arr.constData(), arr.length());
-            socket->flush();
-        }
+        sendClientData(socket, reply);
+    }
+}
+
+void BrowserHost::sendClientMessage(QLocalSocket* socket, const QJsonObject& json)
+{
+    QString reply(QJsonDocument(json).toJson(QJsonDocument::Compact));
+    sendClientData(socket, reply);
+}
+
+void BrowserHost::sendClientData(QLocalSocket* socket, const QString& data)
+{
+    if (socket && socket->isValid() && socket->state() == QLocalSocket::ConnectedState) {
+        QByteArray arr = data.toUtf8();
+        socket->write(arr.constData(), arr.length());
+        socket->flush();
     }
 }
 

--- a/src/browser/BrowserHost.h
+++ b/src/browser/BrowserHost.h
@@ -24,6 +24,7 @@
 
 class QLocalServer;
 class QLocalSocket;
+class QString;
 
 class BrowserHost : public QObject
 {
@@ -36,15 +37,19 @@ public:
     void start();
     void stop();
 
-    void sendClientMessage(const QJsonObject& json);
+    void broadcastClientMessage(const QJsonObject& json);
+    void sendClientMessage(QLocalSocket* socket, const QJsonObject& json);
 
 signals:
-    void clientMessageReceived(const QJsonObject& json);
+    void clientMessageReceived(QLocalSocket* socket, const QJsonObject& json);
 
 private slots:
     void proxyConnected();
     void readProxyMessage();
     void proxyDisconnected();
+
+private:
+    void sendClientData(QLocalSocket* socket, const QString& data);
 
 private:
     QPointer<QLocalServer> m_localServer;

--- a/src/browser/BrowserMessageBuilder.cpp
+++ b/src/browser/BrowserMessageBuilder.cpp
@@ -1,0 +1,259 @@
+/*
+ *  Copyright (C) 2022 KeePassXC Team <team@keepassxc.org>
+ *
+ *  This program is free software: you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation, either version 3 of the License, or
+ *  (at your option) any later version.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "BrowserMessageBuilder.h"
+#include "BrowserShared.h"
+#include "config-keepassx.h"
+#include "core/Global.h"
+#include "core/Tools.h"
+
+#include <QJsonArray>
+#include <QJsonDocument>
+#include <QJsonObject>
+
+#include <botan/sodium.h>
+
+using namespace Botan::Sodium;
+
+Q_GLOBAL_STATIC(BrowserMessageBuilder, s_browserMessageBuilder);
+
+BrowserMessageBuilder* BrowserMessageBuilder::instance()
+{
+    return s_browserMessageBuilder;
+}
+
+QPair<QString, QString> BrowserMessageBuilder::getKeyPair()
+{
+    unsigned char pk[crypto_box_PUBLICKEYBYTES];
+    unsigned char sk[crypto_box_SECRETKEYBYTES];
+    crypto_box_keypair(pk, sk);
+
+    const QString publicKey = getBase64FromKey(pk, crypto_box_PUBLICKEYBYTES);
+    const QString secretKey = getBase64FromKey(sk, crypto_box_SECRETKEYBYTES);
+    return qMakePair(publicKey, secretKey);
+}
+
+QJsonObject BrowserMessageBuilder::getErrorReply(const QString& action, const int errorCode) const
+{
+    QJsonObject response;
+    response["action"] = action;
+    response["errorCode"] = QString::number(errorCode);
+    response["error"] = getErrorMessage(errorCode);
+    return response;
+}
+
+QJsonObject BrowserMessageBuilder::buildMessage(const QString& nonce) const
+{
+    QJsonObject message;
+    message["version"] = KEEPASSXC_VERSION;
+    message["success"] = TRUE_STR;
+    message["nonce"] = nonce;
+    return message;
+}
+
+QJsonObject BrowserMessageBuilder::buildResponse(const QString& action,
+                                                 const QJsonObject& message,
+                                                 const QString& nonce,
+                                                 const QString& publicKey,
+                                                 const QString& secretKey)
+{
+    QJsonObject response;
+    QString encryptedMessage = encryptMessage(message, nonce, publicKey, secretKey);
+    if (encryptedMessage.isEmpty()) {
+        return getErrorReply(action, ERROR_KEEPASS_CANNOT_ENCRYPT_MESSAGE);
+    }
+
+    response["action"] = action;
+    response["message"] = encryptedMessage;
+    response["nonce"] = nonce;
+    return response;
+}
+
+QString BrowserMessageBuilder::getErrorMessage(const int errorCode) const
+{
+    switch (errorCode) {
+    case ERROR_KEEPASS_DATABASE_NOT_OPENED:
+        return QObject::tr("Database not opened");
+    case ERROR_KEEPASS_DATABASE_HASH_NOT_RECEIVED:
+        return QObject::tr("Database hash not available");
+    case ERROR_KEEPASS_CLIENT_PUBLIC_KEY_NOT_RECEIVED:
+        return QObject::tr("Client public key not received");
+    case ERROR_KEEPASS_CANNOT_DECRYPT_MESSAGE:
+        return QObject::tr("Cannot decrypt message");
+    case ERROR_KEEPASS_ACTION_CANCELLED_OR_DENIED:
+        return QObject::tr("Action cancelled or denied");
+    case ERROR_KEEPASS_CANNOT_ENCRYPT_MESSAGE:
+        return QObject::tr("Message encryption failed.");
+    case ERROR_KEEPASS_ASSOCIATION_FAILED:
+        return QObject::tr("KeePassXC association failed, try again");
+    case ERROR_KEEPASS_ENCRYPTION_KEY_UNRECOGNIZED:
+        return QObject::tr("Encryption key is not recognized");
+    case ERROR_KEEPASS_INCORRECT_ACTION:
+        return QObject::tr("Incorrect action");
+    case ERROR_KEEPASS_EMPTY_MESSAGE_RECEIVED:
+        return QObject::tr("Empty message received");
+    case ERROR_KEEPASS_NO_URL_PROVIDED:
+        return QObject::tr("No URL provided");
+    case ERROR_KEEPASS_NO_LOGINS_FOUND:
+        return QObject::tr("No logins found");
+    case ERROR_KEEPASS_NO_GROUPS_FOUND:
+        return QObject::tr("No groups found");
+    case ERROR_KEEPASS_CANNOT_CREATE_NEW_GROUP:
+        return QObject::tr("Cannot create new group");
+    case ERROR_KEEPASS_NO_VALID_UUID_PROVIDED:
+        return QObject::tr("No valid UUID provided");
+    default:
+        return QObject::tr("Unknown error");
+    }
+}
+
+QString BrowserMessageBuilder::encryptMessage(const QJsonObject& message,
+                                              const QString& nonce,
+                                              const QString& publicKey,
+                                              const QString& secretKey)
+{
+    if (message.isEmpty() || nonce.isEmpty()) {
+        return QString();
+    }
+
+    const QString reply(QJsonDocument(message).toJson());
+    if (!reply.isEmpty()) {
+        return encrypt(reply, nonce, publicKey, secretKey);
+    }
+
+    return QString();
+}
+
+QJsonObject BrowserMessageBuilder::decryptMessage(const QString& message,
+                                                  const QString& nonce,
+                                                  const QString& publicKey,
+                                                  const QString& secretKey)
+{
+    if (message.isEmpty() || nonce.isEmpty()) {
+        return QJsonObject();
+    }
+
+    QByteArray ba = decrypt(message, nonce, publicKey, secretKey);
+    if (ba.isEmpty()) {
+        return QJsonObject();
+    }
+
+    return getJsonObject(ba);
+}
+
+QString BrowserMessageBuilder::encrypt(const QString& plaintext,
+                                       const QString& nonce,
+                                       const QString& publicKey,
+                                       const QString& secretKey)
+{
+    const QByteArray ma = plaintext.toUtf8();
+    const QByteArray na = base64Decode(nonce);
+    const QByteArray ca = base64Decode(publicKey);
+    const QByteArray sa = base64Decode(secretKey);
+
+    std::vector<unsigned char> m(ma.cbegin(), ma.cend());
+    std::vector<unsigned char> n(na.cbegin(), na.cend());
+    std::vector<unsigned char> ck(ca.cbegin(), ca.cend());
+    std::vector<unsigned char> sk(sa.cbegin(), sa.cend());
+
+    std::vector<unsigned char> e;
+    e.resize(BrowserShared::NATIVEMSG_MAX_LENGTH);
+
+    if (m.empty() || n.empty() || ck.empty() || sk.empty()) {
+        return QString();
+    }
+
+    if (crypto_box_easy(e.data(), m.data(), m.size(), n.data(), ck.data(), sk.data()) == 0) {
+        QByteArray res = getQByteArray(e.data(), (crypto_box_MACBYTES + ma.length()));
+        return res.toBase64();
+    }
+
+    return QString();
+}
+
+QByteArray BrowserMessageBuilder::decrypt(const QString& encrypted,
+                                          const QString& nonce,
+                                          const QString& publicKey,
+                                          const QString& secretKey)
+{
+    const QByteArray ma = base64Decode(encrypted);
+    const QByteArray na = base64Decode(nonce);
+    const QByteArray ca = base64Decode(publicKey);
+    const QByteArray sa = base64Decode(secretKey);
+
+    std::vector<unsigned char> m(ma.cbegin(), ma.cend());
+    std::vector<unsigned char> n(na.cbegin(), na.cend());
+    std::vector<unsigned char> ck(ca.cbegin(), ca.cend());
+    std::vector<unsigned char> sk(sa.cbegin(), sa.cend());
+
+    std::vector<unsigned char> d;
+    d.resize(BrowserShared::NATIVEMSG_MAX_LENGTH);
+
+    if (m.empty() || n.empty() || ck.empty() || sk.empty()) {
+        return QByteArray();
+    }
+
+    if (crypto_box_open_easy(d.data(), m.data(), ma.length(), n.data(), ck.data(), sk.data()) == 0) {
+        return getQByteArray(d.data(), std::char_traits<char>::length(reinterpret_cast<const char*>(d.data())));
+    }
+
+    return QByteArray();
+}
+
+QString BrowserMessageBuilder::getBase64FromKey(const uchar* array, const uint len)
+{
+    return getQByteArray(array, len).toBase64();
+}
+
+QByteArray BrowserMessageBuilder::getQByteArray(const uchar* array, const uint len) const
+{
+    QByteArray qba;
+    qba.reserve(len);
+    for (uint i = 0; i < len; ++i) {
+        qba.append(static_cast<char>(array[i]));
+    }
+    return qba;
+}
+
+QJsonObject BrowserMessageBuilder::getJsonObject(const uchar* pArray, const uint len) const
+{
+    QByteArray arr = getQByteArray(pArray, len);
+    QJsonParseError err;
+    QJsonDocument doc(QJsonDocument::fromJson(arr, &err));
+    return doc.object();
+}
+
+QJsonObject BrowserMessageBuilder::getJsonObject(const QByteArray& ba) const
+{
+    QJsonParseError err;
+    QJsonDocument doc(QJsonDocument::fromJson(ba, &err));
+    return doc.object();
+}
+
+QByteArray BrowserMessageBuilder::base64Decode(const QString& str)
+{
+    return QByteArray::fromBase64(str.toUtf8());
+}
+
+QString BrowserMessageBuilder::incrementNonce(const QString& nonce)
+{
+    const QByteArray nonceArray = base64Decode(nonce);
+    std::vector<unsigned char> n(nonceArray.cbegin(), nonceArray.cend());
+
+    sodium_increment(n.data(), n.size());
+    return getQByteArray(n.data(), n.size()).toBase64();
+}

--- a/src/browser/BrowserMessageBuilder.h
+++ b/src/browser/BrowserMessageBuilder.h
@@ -1,0 +1,96 @@
+/*
+ *  Copyright (C) 2022 KeePassXC Team <team@keepassxc.org>
+ *
+ *  This program is free software: you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation, either version 3 of the License, or
+ *  (at your option) any later version.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#ifndef BROWSERMESSAGEBUILDER_H
+#define BROWSERMESSAGEBUILDER_H
+
+#include <QPair>
+#include <QString>
+
+class QJsonObject;
+
+namespace
+{
+    enum
+    {
+        ERROR_KEEPASS_DATABASE_NOT_OPENED = 1,
+        ERROR_KEEPASS_DATABASE_HASH_NOT_RECEIVED = 2,
+        ERROR_KEEPASS_CLIENT_PUBLIC_KEY_NOT_RECEIVED = 3,
+        ERROR_KEEPASS_CANNOT_DECRYPT_MESSAGE = 4,
+        ERROR_KEEPASS_TIMEOUT_OR_NOT_CONNECTED = 5,
+        ERROR_KEEPASS_ACTION_CANCELLED_OR_DENIED = 6,
+        ERROR_KEEPASS_CANNOT_ENCRYPT_MESSAGE = 7,
+        ERROR_KEEPASS_ASSOCIATION_FAILED = 8,
+        ERROR_KEEPASS_KEY_CHANGE_FAILED = 9,
+        ERROR_KEEPASS_ENCRYPTION_KEY_UNRECOGNIZED = 10,
+        ERROR_KEEPASS_NO_SAVED_DATABASES_FOUND = 11,
+        ERROR_KEEPASS_INCORRECT_ACTION = 12,
+        ERROR_KEEPASS_EMPTY_MESSAGE_RECEIVED = 13,
+        ERROR_KEEPASS_NO_URL_PROVIDED = 14,
+        ERROR_KEEPASS_NO_LOGINS_FOUND = 15,
+        ERROR_KEEPASS_NO_GROUPS_FOUND = 16,
+        ERROR_KEEPASS_CANNOT_CREATE_NEW_GROUP = 17,
+        ERROR_KEEPASS_NO_VALID_UUID_PROVIDED = 18
+    };
+}
+
+class BrowserMessageBuilder
+{
+public:
+    explicit BrowserMessageBuilder() = default;
+    static BrowserMessageBuilder* instance();
+
+    QPair<QString, QString> getKeyPair();
+
+    QJsonObject buildMessage(const QString& nonce) const;
+    QJsonObject buildResponse(const QString& action,
+                              const QJsonObject& message,
+                              const QString& nonce,
+                              const QString& publicKey,
+                              const QString& secretKey);
+    QJsonObject getErrorReply(const QString& action, const int errorCode) const;
+    QString getErrorMessage(const int errorCode) const;
+
+    QString encryptMessage(const QJsonObject& message,
+                           const QString& nonce,
+                           const QString& publicKey,
+                           const QString& secretKey);
+    QJsonObject
+    decryptMessage(const QString& message, const QString& nonce, const QString& publicKey, const QString& secretKey);
+    QString encrypt(const QString& plaintext, const QString& nonce, const QString& publicKey, const QString& secretKey);
+    QByteArray
+    decrypt(const QString& encrypted, const QString& nonce, const QString& publicKey, const QString& secretKey);
+
+    QString getBase64FromKey(const uchar* array, const uint len);
+    QByteArray getQByteArray(const uchar* array, const uint len) const;
+    QJsonObject getJsonObject(const uchar* pArray, const uint len) const;
+    QJsonObject getJsonObject(const QByteArray& ba) const;
+    QByteArray base64Decode(const QString& str);
+    QString incrementNonce(const QString& nonce);
+
+private:
+    Q_DISABLE_COPY(BrowserMessageBuilder);
+
+    // friend class TestBrowser;
+};
+
+static inline BrowserMessageBuilder* browserMessageBuilder()
+{
+    return BrowserMessageBuilder::instance();
+}
+
+#endif // BROWSERMESSAGEBUILDER_H

--- a/src/browser/BrowserMessageBuilder.h
+++ b/src/browser/BrowserMessageBuilder.h
@@ -85,7 +85,7 @@ public:
 private:
     Q_DISABLE_COPY(BrowserMessageBuilder);
 
-    // friend class TestBrowser;
+    friend class TestBrowser;
 };
 
 static inline BrowserMessageBuilder* browserMessageBuilder()

--- a/src/browser/BrowserService.cpp
+++ b/src/browser/BrowserService.cpp
@@ -23,6 +23,7 @@
 #include "BrowserEntryConfig.h"
 #include "BrowserEntrySaveDialog.h"
 #include "BrowserHost.h"
+#include "BrowserMessageBuilder.h"
 #include "BrowserSettings.h"
 #include "core/Tools.h"
 #include "gui/MainWindow.h"
@@ -64,6 +65,7 @@ BrowserService::BrowserService()
     , m_browserHost(new BrowserHost)
     , m_dialogActive(false)
     , m_bringToFrontRequested(false)
+    , m_passwordGeneratorRequested(false)
     , m_prevWindowState(WindowState::Normal)
     , m_keepassBrowserUUID(Tools::hexToUuid("de887cc3036343b8974b5911b8816224"))
 {
@@ -311,26 +313,37 @@ QString BrowserService::getCurrentTotp(const QString& uuid)
     return {};
 }
 
-void BrowserService::showPasswordGenerator(const QJsonObject& errorMessage, const QString& nonce)
+void BrowserService::showPasswordGenerator(const QString& incrementedNonce,
+                                           const QString& publicKey,
+                                           const QString& secretKey)
 {
     if (!m_passwordGenerator) {
         m_passwordGenerator.reset(PasswordGeneratorWidget::popupGenerator());
 
         connect(m_passwordGenerator.data(), &PasswordGeneratorWidget::closed, m_passwordGenerator.data(), [=] {
             if (!m_passwordGenerator->isPasswordGenerated()) {
+                auto errorMessage = browserMessageBuilder()->getErrorReply("generate-password",
+                                                                           ERROR_KEEPASS_ACTION_CANCELLED_OR_DENIED);
                 m_browserHost->sendClientMessage(errorMessage);
             }
 
             m_passwordGenerator.reset();
             hideWindow();
+            m_passwordGeneratorRequested = false;
         });
 
         connect(m_passwordGenerator.data(),
                 &PasswordGeneratorWidget::appliedPassword,
                 m_passwordGenerator.data(),
-                [=](const QString& password) { emit passwordGenerated(password, nonce); });
+                [=](const QString& password) {
+                    QJsonObject message = browserMessageBuilder()->buildMessage(incrementedNonce);
+                    message["password"] = password;
+                    sendPassword(browserMessageBuilder()->buildResponse(
+                        "generate-password", message, incrementedNonce, publicKey, secretKey));
+                });
     }
 
+    m_passwordGeneratorRequested = true;
     raiseWindow();
     m_passwordGenerator->raise();
     m_passwordGenerator->activateWindow();
@@ -340,6 +353,11 @@ void BrowserService::sendPassword(const QJsonObject& message)
 {
     m_browserHost->sendClientMessage(message);
     hideWindow();
+}
+
+bool BrowserService::isPasswordGeneratorRequested() const
+{
+    return m_passwordGeneratorRequested;
 }
 
 QString BrowserService::storeKey(const QString& key)

--- a/src/browser/BrowserService.cpp
+++ b/src/browser/BrowserService.cpp
@@ -40,6 +40,7 @@
 #include <QJsonArray>
 #include <QJsonObject>
 #include <QListWidget>
+#include <QLocalSocket>
 #include <QProgressDialog>
 #include <QUrl>
 
@@ -313,7 +314,8 @@ QString BrowserService::getCurrentTotp(const QString& uuid)
     return {};
 }
 
-void BrowserService::showPasswordGenerator(const QString& incrementedNonce,
+void BrowserService::showPasswordGenerator(QLocalSocket* socket,
+                                           const QString& incrementedNonce,
                                            const QString& publicKey,
                                            const QString& secretKey)
 {
@@ -324,7 +326,7 @@ void BrowserService::showPasswordGenerator(const QString& incrementedNonce,
             if (!m_passwordGenerator->isPasswordGenerated()) {
                 auto errorMessage = browserMessageBuilder()->getErrorReply("generate-password",
                                                                            ERROR_KEEPASS_ACTION_CANCELLED_OR_DENIED);
-                m_browserHost->sendClientMessage(errorMessage);
+                m_browserHost->sendClientMessage(socket, errorMessage);
             }
 
             m_passwordGenerator.reset();
@@ -338,8 +340,9 @@ void BrowserService::showPasswordGenerator(const QString& incrementedNonce,
                 [=](const QString& password) {
                     QJsonObject message = browserMessageBuilder()->buildMessage(incrementedNonce);
                     message["password"] = password;
-                    sendPassword(browserMessageBuilder()->buildResponse(
-                        "generate-password", message, incrementedNonce, publicKey, secretKey));
+                    sendPassword(socket,
+                                 browserMessageBuilder()->buildResponse(
+                                     "generate-password", message, incrementedNonce, publicKey, secretKey));
                 });
     }
 
@@ -349,9 +352,9 @@ void BrowserService::showPasswordGenerator(const QString& incrementedNonce,
     m_passwordGenerator->activateWindow();
 }
 
-void BrowserService::sendPassword(const QJsonObject& message)
+void BrowserService::sendPassword(QLocalSocket* socket, const QJsonObject& message)
 {
-    m_browserHost->sendClientMessage(message);
+    m_browserHost->sendClientMessage(socket, message);
     hideWindow();
 }
 
@@ -1400,7 +1403,7 @@ void BrowserService::databaseLocked(DatabaseWidget* dbWidget)
     if (dbWidget) {
         QJsonObject msg;
         msg["action"] = QString("database-locked");
-        m_browserHost->sendClientMessage(msg);
+        m_browserHost->broadcastClientMessage(msg);
     }
 }
 
@@ -1414,7 +1417,7 @@ void BrowserService::databaseUnlocked(DatabaseWidget* dbWidget)
 
         QJsonObject msg;
         msg["action"] = QString("database-unlocked");
-        m_browserHost->sendClientMessage(msg);
+        m_browserHost->broadcastClientMessage(msg);
 
         auto db = dbWidget->database();
         if (checkLegacySettings(db)) {
@@ -1437,7 +1440,7 @@ void BrowserService::activeDatabaseChanged(DatabaseWidget* dbWidget)
     m_currentDatabaseWidget = dbWidget;
 }
 
-void BrowserService::processClientMessage(const QJsonObject& message)
+void BrowserService::processClientMessage(QLocalSocket* socket, const QJsonObject& message)
 {
     auto clientID = message["clientID"].toString();
     if (clientID.isEmpty()) {
@@ -1450,6 +1453,6 @@ void BrowserService::processClientMessage(const QJsonObject& message)
     }
 
     auto& action = m_browserClients.value(clientID);
-    auto response = action->processClientMessage(message);
-    m_browserHost->sendClientMessage(response);
+    auto response = action->processClientMessage(socket, message);
+    m_browserHost->sendClientMessage(socket, response);
 }

--- a/src/browser/BrowserService.h
+++ b/src/browser/BrowserService.h
@@ -23,6 +23,8 @@
 #include "core/Entry.h"
 #include "gui/PasswordGeneratorWidget.h"
 
+class QLocalSocket;
+
 typedef QPair<QString, QString> StringPair;
 typedef QList<StringPair> StringPairList;
 
@@ -56,8 +58,11 @@ public:
     QJsonObject getDatabaseGroups();
     QJsonObject createNewGroup(const QString& groupName);
     QString getCurrentTotp(const QString& uuid);
-    void showPasswordGenerator(const QString& nonce, const QString& publicKey, const QString& secretKey);
-    void sendPassword(const QJsonObject& message);
+    void showPasswordGenerator(QLocalSocket* socket,
+                               const QString& nonce,
+                               const QString& publicKey,
+                               const QString& secretKey);
+    void sendPassword(QLocalSocket* socket, const QJsonObject& message);
     bool isPasswordGeneratorRequested() const;
 
     void addEntry(const QString& dbid,
@@ -98,7 +103,7 @@ public:
 
 signals:
     void requestUnlock();
-    void passwordGenerated(const QString& password, const QString& nonce);
+    void passwordGenerated(QLocalSocket* socket, const QString& password, const QString& nonce);
 
 public slots:
     void databaseLocked(DatabaseWidget* dbWidget);
@@ -106,7 +111,7 @@ public slots:
     void activeDatabaseChanged(DatabaseWidget* dbWidget);
 
 private slots:
-    void processClientMessage(const QJsonObject& message);
+    void processClientMessage(QLocalSocket* socket, const QJsonObject& message);
 
 private:
     enum Access

--- a/src/browser/BrowserService.h
+++ b/src/browser/BrowserService.h
@@ -56,8 +56,9 @@ public:
     QJsonObject getDatabaseGroups();
     QJsonObject createNewGroup(const QString& groupName);
     QString getCurrentTotp(const QString& uuid);
-    void showPasswordGenerator(const QJsonObject& errorMessage, const QString& nonce);
+    void showPasswordGenerator(const QString& nonce, const QString& publicKey, const QString& secretKey);
     void sendPassword(const QJsonObject& message);
+    bool isPasswordGeneratorRequested() const;
 
     void addEntry(const QString& dbid,
                   const QString& login,
@@ -163,6 +164,7 @@ private:
 
     bool m_dialogActive;
     bool m_bringToFrontRequested;
+    bool m_passwordGeneratorRequested;
     WindowState m_prevWindowState;
     QUuid m_keepassBrowserUUID;
 

--- a/src/browser/CMakeLists.txt
+++ b/src/browser/CMakeLists.txt
@@ -1,5 +1,5 @@
 #  Copyright (C) 2017 Sami Vänttinen <sami.vanttinen@protonmail.com>
-#  Copyright (C) 2017 KeePassXC Team <team@keepassxc.org>
+#  Copyright (C) 2022 KeePassXC Team <team@keepassxc.org>
 #
 #  This program is free software: you can redistribute it and/or modify
 #  it under the terms of the GNU General Public License as published by
@@ -23,6 +23,7 @@ if(WITH_XC_BROWSER)
             BrowserEntryConfig.cpp
             BrowserEntrySaveDialog.cpp
             BrowserHost.cpp
+            BrowserMessageBuilder.cpp
             BrowserSettingsPage.cpp
             BrowserSettingsWidget.cpp
             BrowserService.cpp

--- a/tests/TestBrowser.cpp
+++ b/tests/TestBrowser.cpp
@@ -17,6 +17,7 @@
 
 #include "TestBrowser.h"
 
+#include "browser/BrowserMessageBuilder.h"
 #include "browser/BrowserSettings.h"
 #include "core/Group.h"
 #include "core/Tools.h"
@@ -75,7 +76,7 @@ void TestBrowser::testEncryptMessage()
     m_browserAction->m_publicKey = SERVERPUBLICKEY;
     m_browserAction->m_secretKey = SERVERSECRETKEY;
     m_browserAction->m_clientPublicKey = PUBLICKEY;
-    auto encrypted = m_browserAction->encryptMessage(message, NONCE);
+    auto encrypted = browserMessageBuilder()->encryptMessage(message, NONCE, PUBLICKEY, SERVERSECRETKEY);
 
     QCOMPARE(encrypted, QString("+zjtntnk4rGWSl/Ph7Vqip/swvgeupk4lNgHEm2OO3ujNr0OMz6eQtGwjtsj+/rP"));
 }
@@ -86,7 +87,7 @@ void TestBrowser::testDecryptMessage()
     m_browserAction->m_publicKey = SERVERPUBLICKEY;
     m_browserAction->m_secretKey = SERVERSECRETKEY;
     m_browserAction->m_clientPublicKey = PUBLICKEY;
-    auto decrypted = m_browserAction->decryptMessage(message, NONCE);
+    auto decrypted = browserMessageBuilder()->decryptMessage(message, NONCE, PUBLICKEY, SERVERSECRETKEY);
 
     QCOMPARE(decrypted["action"].toString(), QString("test-action"));
 }
@@ -99,13 +100,13 @@ void TestBrowser::testGetBase64FromKey()
         pk[i] = i;
     }
 
-    auto response = m_browserAction->getBase64FromKey(pk, crypto_box_PUBLICKEYBYTES);
+    auto response = browserMessageBuilder()->getBase64FromKey(pk, crypto_box_PUBLICKEYBYTES);
     QCOMPARE(response, QString("AAECAwQFBgcICQoLDA0ODxAREhMUFRYXGBkaGxwdHh8="));
 }
 
 void TestBrowser::testIncrementNonce()
 {
-    auto result = m_browserAction->incrementNonce(NONCE);
+    auto result = browserMessageBuilder()->incrementNonce(NONCE);
     QCOMPARE(result, QString("zRKdvTjL5bgWaKMCTut/8soM/uoMrFoZ"));
 }
 

--- a/tests/TestBrowser.cpp
+++ b/tests/TestBrowser.cpp
@@ -62,7 +62,7 @@ void TestBrowser::testChangePublicKeys()
     json["publicKey"] = PUBLICKEY;
     json["nonce"] = NONCE;
 
-    auto response = m_browserAction->processClientMessage(json);
+    auto response = m_browserAction->processClientMessage(nullptr, json);
     QCOMPARE(response["action"].toString(), QString("change-public-keys"));
     QCOMPARE(response["publicKey"].toString() == PUBLICKEY, false);
     QCOMPARE(response["success"].toString(), TRUE_STR);

